<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mark Dunning">
<meta name="dcterms.date" content="2025-11-27">
<meta name="description" content="Optimising Machine Leaerning Performance by cross-validation and tuning">

<title>Tidymodels for omics data: Part 4 – Dr.&nbsp;Mark Dunning - Bioinformatician / Data Scientist</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-2fef5ea3f8957b3e4ecc936fc74692ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-1688d9e1630bd0bcb1ec498477fd8fb7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-96TB60ETKS"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-96TB60ETKS', { 'anonymize_ip': true});
</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Dr.&nbsp;Mark Dunning - Bioinformatician / Data Scientist</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../cv.html"> 
<span class="menu-text">CV</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../training.html"> 
<span class="menu-text">Training Materials</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../posts.html"> 
<span class="menu-text">Posts</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#pre-amble" id="toc-pre-amble" class="nav-link active" data-scroll-target="#pre-amble">Pre-amble</a></li>
  <li><a href="#cross-validation" id="toc-cross-validation" class="nav-link" data-scroll-target="#cross-validation">Cross-Validation</a></li>
  <li><a href="#model-specifiction-and-recipe" id="toc-model-specifiction-and-recipe" class="nav-link" data-scroll-target="#model-specifiction-and-recipe">Model specifiction and recipe</a></li>
  <li><a href="#setup-a-grid-for-tuning" id="toc-setup-a-grid-for-tuning" class="nav-link" data-scroll-target="#setup-a-grid-for-tuning">Setup a grid for tuning</a></li>
  <li><a href="#inspect-tuning-results" id="toc-inspect-tuning-results" class="nav-link" data-scroll-target="#inspect-tuning-results">Inspect tuning results</a></li>
  <li><a href="#create-final-model-and-fit" id="toc-create-final-model-and-fit" class="nav-link" data-scroll-target="#create-final-model-and-fit">Create final model and fit</a></li>
  <li><a href="#training-for-random-forests" id="toc-training-for-random-forests" class="nav-link" data-scroll-target="#training-for-random-forests">Training for Random Forests</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Tidymodels for omics data: Part 4</h1>
</div>

<div>
  <div class="description">
    Optimising Machine Leaerning Performance by cross-validation and tuning
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mark Dunning </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">November 27, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="pre-amble" class="level2">
<h2 class="anchored" data-anchor-id="pre-amble">Pre-amble</h2>
<p>In <a href="https://mdbioinformatics.com/posts/2025_11_06_tidymodels_TCGA_part1/index.html">Part 1 of this series</a> we described how to download TCGA data for breast cancers and manipulated them using a combination of <code>tidybulk</code> and <code>dplyr</code> to retain a set of expressed, variable genes. There are a set of packages that we will need for this Part 3b:-</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(tidymodels)) <span class="fu">install.packages</span>(<span class="st">"tidymodels"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(dplyr)) <span class="fu">install.packages</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(ggplot2)) <span class="fu">install.packages</span>(<span class="st">"ggplot2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You will need the processed data from the <a href="https://mdbioinformatics.com/posts/2025_11_06_tidymodels_TCGA_part1/">Part one of this series</a> and the code to download this is:-</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">### get the saved RDS</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"raw_data"</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">"raw_data/brca_train_tidy.rds"</span>)) <span class="fu">download.file</span>(<span class="st">"https://github.com/markdunning/markdunning.github.com/raw/refs/heads/master/posts/2025_11_06_tidymodels_TCGA_part1/brca_train_tidy.rds"</span>, <span class="at">destfile =</span> <span class="st">"raw_data/brca_train_tidy.rds"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">"raw_data/brca_test_tidy.rds"</span>)) <span class="fu">download.file</span>(<span class="st">"https://github.com/markdunning/markdunning.github.com/raw/refs/heads/master/posts/2025_11_06_tidymodels_TCGA_part1/brca_test_tidy.rds"</span>, <span class="at">destfile =</span> <span class="st">"raw_data/brca_test_tidy.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The main purpose of this Part of the series is to introduce some other classification methods before proceeding to more complex topics to as tuning our models. First we will load the pre-prepared tidy data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>brca_train_tidy <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"raw_data/brca_train_tidy.rds"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>brca_test_tidy <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"raw_data/brca_test_tidy.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s proceed by picking genes with an <em>absolute</em> correlation &lt; 0.7 with ESR1 as our possible predictors.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>kept_genes <span class="ot">&lt;-</span> brca_train_tidy <span class="sc">%&gt;%</span> </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(CLEC3A<span class="sc">:</span>BRINP2) <span class="sc">%&gt;%</span> </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cor</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"Cor_with_Gene"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(ESR1, Cor_with_Gene) <span class="sc">%&gt;%</span> <span class="co">#</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">abs</span>(ESR1) <span class="sc">&lt;</span><span class="fl">0.7</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(Cor_with_Gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we restrict our data to just these genes, and the Estrogen Receptor status (renamed to <code>ER</code> for convenience). We will also create some ID rows in the data, which will be helpful in a short while.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>er_train <span class="ot">&lt;-</span> brca_train_tidy <span class="sc">%&gt;%</span> </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">all_of</span>(kept_genes), <span class="at">ER =</span> er_status_by_ihc) <span class="sc">%&gt;%</span> </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ER =</span> <span class="fu">as.factor</span>(ER)) <span class="sc">%&gt;%</span> </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ID =</span> <span class="fu">paste0</span>(<span class="st">"Training_"</span>, <span class="fu">row_number</span>())) <span class="sc">%&gt;%</span> </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(ID)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>er_test <span class="ot">&lt;-</span> brca_test_tidy <span class="sc">%&gt;%</span> </span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">all_of</span>(kept_genes), <span class="at">ER =</span> er_status_by_ihc) <span class="sc">%&gt;%</span> </span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ER =</span> <span class="fu">as.factor</span>(ER)) <span class="sc">%&gt;%</span> </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ID =</span> <span class="fu">paste0</span>(<span class="st">"Training_"</span>, <span class="fu">row_number</span>())) <span class="sc">%&gt;%</span> </span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(ID)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In <a href="https://mdbioinformatics.com/posts/2025_11_17_tidymodels_TCGA_part3/#introducing-glmnet">Part 3a</a> we introduced some machine learning methods that relied on hyperparameters in order to be run. This included logistic regression using glmnet that can “shrink” the influence of features in our data. The amount of shrinkage is controlled by the arguments <code>penalty</code> (<span class="math inline">\(\lambda\)</span>) and <code>mixture</code> (<span class="math inline">\(\alpha\)</span>) in the tidymodels specification.</p>
<p>We just set these to some arbitrary, fixed values. So, the question is: How can we choose the “best” values for these parameters without endless trial and error, and what data should we base that decision on?</p>
<p>Way back in <a href="https://mdbioinformatics.com/posts/2025_11_06_tidymodels_TCGA_part1/#preparing-the-data-for-machine-learning">Part 1</a> we split a TCGA breast cancer dataset into training and testing portions that were completely independent. The roles of these two portions are strictly defined:</p>
<ul>
<li><p>Training Dataset: Used only to estimate the internal coefficients of our model and create the final fit. In our example of a logsitic regression this would be the amount to which each gene affects the log-odds of a particular sample being ER <code>Positive</code>.</p></li>
<li><p>Testing Dataset: Used only for the final, independent validation of the model’s performance on completely unseen data.</p></li>
</ul>
<p>Neither of these is appropriate for tuning. You might think the training data would be a good choice, but using it directly to pick the hyperparameters would make our model overfit. It would be optimised for the training data and potentially perform poorly on the test set or other data.</p>
<p>So the solution is to create some new, temporary validation sets, based on our original training data, solely for the purpose of identifying (or “tuning”) the best hyperparameters. This technique is known as Cross-Validation (CV).</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expectation setting
</div>
</div>
<div class="callout-body-container callout-body">
<p>The purpose of tuning our models is really just that—fine-tuning to get the absolute best performance. It will not suddenly elevate performance metrics (like <span class="math inline">\(\text{ROC AUC}\)</span> or Accuracy) from around <span class="math inline">\(0.6\)</span> to <span class="math inline">\(0.9\)</span>. Tuning should only be used in circumstances where you are already satisfied with the fundamental performance of your model and just require some minor, incremental adjustments. The performance gains from tuning are typically small (e.g.&nbsp;).</p>
<p>That is to say that you should consider other choices of model, and even the fundamental data processing and feature engineering first, before trying to “tune” a model that is performing sub-par.</p>
<p><strong>The greatest performance gains come from better data and model choice; tuning only optimizes the model you already have.</strong></p>
</div>
</div>
</section>
<section id="cross-validation" class="level2">
<h2 class="anchored" data-anchor-id="cross-validation">Cross-Validation</h2>
<p>K-fold cross-validation (which <code>tidymodels</code> refers to as V-fold cross-validation) splits data into <strong>v</strong> equally-sized partitions to be used to tune hyper-parameters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>er_folds <span class="ot">&lt;-</span> er_train <span class="sc">%&gt;%</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">vfold_cv</span>(<span class="at">v =</span> <span class="dv">10</span>, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">strata =</span> ER)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Each of our <code>v</code> splits comprises a training and testing set, which can be accessed using the same <code>training</code> and <code>testing</code> functions from before.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>er_folds<span class="sc">$</span>splits[[<span class="dv">1</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;Analysis/Assess/Total&gt;
&lt;761/86/847&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>er_folds<span class="sc">$</span>splits[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> <span class="fu">training</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            ID    CLEC3A   SCGB2A2     CPB1     GSTM1      TFF1    S100A7
1   Training_1  5.130258  7.531285 7.361903  2.654743  1.974253  4.217648
2   Training_2 10.880983 14.974967 5.176676  7.004891 12.640878  8.103595
3   Training_3  1.974253 10.748514 4.059635  7.631561  8.218780  7.752409
4   Training_4  5.962898 12.580000 4.015388 12.110432  8.843665  1.974253
5   Training_5  2.717920  6.267474 1.974253 10.090312  3.236028  1.974253
6   Training_6  2.614033  4.950619 3.224804  1.974253  2.614033  3.065301
7   Training_7  2.622193  6.071904 3.078798  6.565902  1.974253  3.895366
8   Training_8  1.974253  5.501218 6.867097  9.812725  4.547686 12.956422
9   Training_9  2.685000  1.974253 1.974253  9.618968  1.974253 11.405191
10 Training_10  7.775900 17.372097 3.203721  7.853590  5.598493  7.171189
     SCGB1D2    KCNJ3 LINC00993   VSTM2A  ANKRD30A DSCAM.AS1 AC093001.1
1   7.039168 1.974253  2.654743 1.974253  2.928090  1.974253   2.654743
2  12.741020 6.796671 11.461740 6.887116 11.483260  9.184524   8.936326
3   9.274851 3.582869  1.974253 1.974253  1.974253 11.940238  13.599490
4  10.541767 6.661440 11.933112 6.548917 10.990789  9.567059  11.511148
5   4.466175 1.974253  4.895023 2.717920  2.717920  3.236028   3.417207
6   3.224804 1.974253  1.974253 1.974253  1.974253  1.974253   3.873816
7   4.862395 1.974253  3.078798 3.078798  3.240058  1.974253   2.883191
8   2.721976 2.721976  5.009119 3.020490  3.020490  1.974253   2.721976
9   1.974253 1.974253  1.974253 2.685000  1.974253  1.974253   1.974253
10 12.603865 3.843955 11.616168 7.287774 10.087775  3.758805   4.069371
         PIP  SERPINA6     MUCL1     PRAME    PSPHP1   CYP2B7P     MUC6
1   5.337199  4.064487  4.354854 13.159398  1.974253  5.702875 4.288019
2  10.727447  2.586544  8.459198 10.791420  1.974253  8.061529 2.586544
3  17.037551  2.723064 15.963232  4.244237  8.634898  5.869687 6.473676
4   5.066784 11.252797  7.461692  3.102392  1.974253  4.869677 5.206674
5   3.417207  3.014929  6.206055 11.004211  9.758224  4.047086 3.709572
6   1.974253  4.464880  3.954451 12.170767  1.974253  5.058121 7.308983
7   4.593615  2.883191 10.726162  7.161416  5.281486  4.053116 5.250955
8   4.959155  3.953879 10.936065 11.189095 10.124488  4.056983 4.853518
9   6.368073  1.974253  2.969731 11.756726  1.974253  3.356668 3.356668
10 11.175291  1.974253  6.408641  3.843955 10.440639 12.581035 3.459548
     UGT2B11      CST9    HMGCS2  SLC30A8    TFAP2B    CASP14    HOXB13
1   1.974253  1.974253  1.974253 3.688717  7.638377  3.980538  8.783371
2   3.019723  5.147375  9.449080 8.519761  7.283145  6.628528  7.221934
3  13.713047  1.974253  5.504823 3.582869 14.778956  3.844086  1.974253
4   2.902907 12.885022 10.467307 6.180308 13.016402  2.636477  7.702556
5   1.974253  3.832517  2.717920 2.717920  4.996580  3.572606 12.226666
6   1.974253  1.974253  3.065301 3.362422  1.974253  3.065301  1.974253
7   1.974253  1.974253  4.320579 4.734535  5.674644  4.258627  2.883191
8   1.974253  1.974253  1.974253 3.242643  5.355426  6.134179  4.056983
9   1.974253  1.974253  1.974253 1.974253  2.685000  4.686374  8.846480
10  2.602776  1.974253  4.069371 9.994421 13.457986 10.232898 11.041156
      CYP4Z1 AL157387.1       GP2 AC093838.1   BRINP3     FDCSP    CALML5
1   4.064487   1.974253  1.974253   1.974253 3.132496  6.783032  3.793839
2   7.692017   9.406949  9.917294   9.648220 1.974253  9.133762  4.957441
3  10.515529   1.974253 11.483574   6.138015 6.419860  5.896268  7.973624
4  11.826037   9.054957  7.857381  10.292254 1.974253  6.873507  8.213311
5   1.974253   2.717920  4.047086   1.974253 1.974253 17.885986 14.347999
6   1.974253   1.974253  2.871920   3.694884 2.871920  4.410821 11.850988
7   4.053116   1.974253  2.622193   1.974253 3.379129  8.356586  5.604233
8   4.152451   2.721976  3.020490   2.721976 5.601513  8.468704 13.731490
9   6.262044   1.974253  2.685000   1.974253 3.638975 17.815788  7.198754
10 12.348479   8.841153  3.758805   1.974253 3.843955  4.623986  6.124400
      ADIPOQ    SLC5A8    GRIA2     KLK11     ADH1B    KCNC2     DHRS2
1   7.281557 13.810066 3.132496 11.608248  8.069971 1.974253  5.775418
2   9.067689  3.627195 3.627195  4.022179 10.825518 3.019723  7.604154
3   8.126775  7.315722 1.974253  9.269799  8.798176 3.244417 14.907694
4  11.884109  1.974253 1.974253  3.266704 12.600108 1.974253 16.176586
5   8.550709  1.974253 1.974253  7.673682  9.550084 1.974253  4.230916
6   6.106450  6.638119 5.280959  8.546295  6.247756 1.974253  4.464880
7   5.219749  8.482224 3.078798 12.257402  6.156102 1.974253  5.121760
8   3.020490  8.428589 4.614534  9.113626  2.721976 1.974253  6.199389
9   2.969731  2.969731 3.356668  4.629385  3.757984 2.685000  4.374161
10  2.602776 10.809253 3.843955  2.602776  2.856360 2.602776  4.529855
       FABP7    CYP4F8     SOX10 AC093297.1     KRT14 LINC02224   CEACAM5
1  10.768478  2.928090 14.555617   3.132496 16.837633  3.300671  6.711183
2   5.487770  3.878902  8.531014   5.577955 11.888952  8.547731 13.954088
3  12.957177 13.370924  3.426630   1.974253  3.244417  1.974253  6.686075
4   3.932899  5.609578  7.482241   2.902907 10.605567  1.974253  4.826679
5   7.108157  2.717920 13.198709   1.974253 11.854920  1.974253  3.709572
6   1.974253  2.614033 13.648626   1.974253  8.757425  1.974253  4.030265
7   7.034015  4.778479 13.748117   1.974253 12.083724  1.974253  2.622193
8  11.553180  3.242643 12.961897   1.974253  9.450523  1.974253  4.056983
9  12.815151  1.974253 13.227309   1.974253  6.065810  1.974253  2.685000
10  3.459548  3.459548  3.339321   1.974253  6.567181  1.974253  3.923636
   AC061961.1      CGA    NXPH1      TFF3     SYT13 AC044784.1 LINC01087
1    1.974253 4.354854 1.974253  3.573460  9.974574   1.974253  1.974253
2    3.717027 9.707688 8.610075 12.695182  5.818601   9.010554  9.391599
3    1.974253 1.974253 5.012477  9.783699  3.021982   6.591925  4.910634
4    3.646549 1.974253 1.974253 12.641787  3.102392   2.902907 10.122728
5    1.974253 1.974253 1.974253  6.769726  1.974253   1.974253  3.014929
6    1.974253 2.871920 1.974253  3.362422  9.094501   1.974253  3.594375
7    1.974253 1.974253 2.622193  3.240058  3.714841   1.974253  2.622193
8    1.974253 1.974253 2.721976  6.770447  7.912427   2.721976  2.721976
9    1.974253 2.969731 1.974253  1.974253  9.727281   1.974253  1.974253
10   2.602776 2.602776 9.059696 11.528234 11.883932   5.940151  3.203721
        DCD     GLRA3     OBP2B      TAT    MYBPC1 LINC01291 SLC25A24P1
1  5.702875  2.928090  4.217648 3.890663  4.288019  8.616629   2.654743
2  2.586544  2.586544  3.305897 5.487770  4.887935  9.056052   3.952564
3  6.986559  3.244417  3.720539 9.186475 13.821662  8.229229   3.021982
4  1.974253 14.882973  3.266704 4.951885  7.908880  9.587742   1.974253
5  1.974253  1.974253  9.864462 3.832517  6.783787  1.974253   6.865388
6  1.974253  2.614033  9.675350 3.787612  4.354503  5.421026   3.954451
7  3.078798  1.974253  9.332146 3.379129  8.298673  9.021308   4.258627
8  3.718221  2.721976  7.755499 4.324785  6.571666  3.424637   3.020490
9  1.974253  2.685000  3.966108 3.966108  5.443218  8.650782   5.788188
10 6.329118  1.974253 11.419905 5.552336  2.856360  8.951251   7.785843
       BEX1     KRT6B    COL2A1     STAC2    CRABP1    CRISP3     OLFM4
1  3.132496 11.494855 15.809452 11.674992  4.418512  3.132496  3.793839
2  2.833902 11.189557  6.376915  8.963573  7.798309  7.646169  7.186770
3  3.426630  2.723064  2.723064 14.583405  4.550743 13.622581 12.828898
4  2.636477  7.555178  3.932899  6.809852  3.533619  4.991241  5.774997
5  3.417207 10.483524  4.466175  7.015901 11.688076  3.014929  7.015901
6  3.065301 14.611489 13.233529 10.646678 11.898066  3.694884  9.444375
7  3.078798 12.561612  6.979721  6.156102 10.005903  3.240058  4.821071
8  3.242643 14.384055  9.526388 13.397413  9.823076  9.536907  6.434106
9  3.182196 10.743710  2.685000 13.703396 11.859286  2.685000  9.100329
10 4.261197  5.178814  3.203721  6.301601 11.891720  8.935709  1.974253
      SRARP ARHGAP40    PVALB   SIRLNT     MUC5B     GABRP     NPY1R    CBLN2
1  1.974253 8.906288 4.479301 2.654743 12.722563 15.710088  5.727468 4.593276
2  8.765831 3.952564 9.859333 4.737546 10.566580 12.214868  9.109572 4.737546
3  9.496444 2.723064 1.974253 4.910634  9.968075  5.469740  3.844086 1.974253
4  8.651489 8.930933 3.102392 2.636477  4.092886  9.635900 14.662466 3.266704
5  1.974253 6.556903 4.142232 1.974253 11.630958 14.749770  7.903351 5.555282
6  5.058121 6.789404 4.295713 1.974253 10.950827 15.337576  7.605260 6.693993
7  1.974253 3.379129 2.622193 1.974253 12.575204 15.350008  9.291900 4.642212
8  2.721976 9.182243 3.718221 1.974253 14.028051 15.800972 12.669537 6.111766
9  1.974253 6.065810 4.225778 1.974253 12.592306 16.858622 11.140610 2.685000
10 2.602776 6.914308 7.356123 2.602776  6.124400  5.377849 10.117598 4.577731
        KRT5     CLCA2   S100A7A    HOXC12       LTF     KRT6A    BMPR1B
1  13.140173  4.646885  3.132496  3.132496  7.524320  8.768717  7.960733
2  12.674879  9.312299  4.655739  6.530857 11.208515  7.360913  6.680054
3   3.956435 13.012844  5.012477  3.021982 14.466858  3.582869  6.788653
4  10.830487  5.532350  1.974253 12.064064  9.533007  3.533619  7.862615
5  15.482282  2.717920  2.717920  3.417207 12.084282 10.137557  8.270552
6  11.164855  3.362422  2.614033  9.518867  5.572675  8.527991  5.814631
7   8.370709  3.976682  2.622193 10.478113 11.319854  5.187837 10.224783
8  13.462865  8.634616  8.065614  6.397555 16.828242  8.100078  5.918597
9  13.724397  5.597224 10.040223  2.685000 17.805787  8.296161  3.638975
10  7.201231 12.348689  4.480234  2.602776 12.083489  4.946952 12.729406
   SERPINA11      KLK5    PYDC1     TCN1    TRARG1 C10orf82 IGHV7.4.1
1   5.677851 15.290601 1.974253 5.798802  5.702875 2.654743  2.928090
2   5.703347  9.280984 3.173241 8.906396  7.172461 9.845880  7.256261
3   1.974253  4.480395 5.896268 3.582869  6.623998 1.974253  5.320028
4  11.678836  5.752508 3.102392 6.354240 10.428144 6.973717 10.934762
5   3.572606  7.173668 8.962360 4.666514  7.205338 3.709572  3.417207
6   2.614033  5.731318 1.974253 2.614033  4.354503 2.871920  6.869209
7   1.974253 11.010012 1.974253 3.379129  4.193638 2.622193  3.808408
8   4.152451 14.158093 3.242643 8.451013  2.721976 2.721976 15.099101
9   1.974253  8.125660 7.780860 1.974253  3.356668 1.974253  6.574030
10  8.446054  3.459548 7.342710 2.602776  1.974253 8.644852 11.736048
   AC008268.1     KIF1A     CST5    ASCL1      DSG3   BPIFB2 IGHV1.69.2
1    3.132496  8.031199 2.654743 2.654743 15.920897 2.928090   1.974253
2    6.299889  4.737546 1.974253 5.440437 10.383948 5.340752   1.974253
3    6.069676  9.289899 1.974253 1.974253  3.582869 7.023074   3.244417
4    4.015388  3.408295 1.974253 2.902907  7.809406 2.902907   4.092886
5    2.717920  8.014741 3.236028 4.230916 12.695906 1.974253   7.732809
6    2.614033 12.004769 1.974253 2.614033  1.974253 3.484371   4.615230
7    1.974253  9.194045 1.974253 2.622193  5.219749 3.078798  14.473058
8    1.974253 11.446754 1.974253 3.424637  7.188864 1.974253   8.286193
9    1.974253  8.444595 1.974253 3.638975  9.050891 1.974253   5.862806
10   7.273707  6.471704 2.602776 2.856360  4.428728 7.589778   4.069371
     CEACAM6   OAZ1P1   SLC34A2   LRRC31   CDC20B      ELF5 LINC00052
1   6.622555 1.974253 11.852935 2.928090 3.445442 14.162286  5.092608
2  17.335894 6.824406  8.664736 4.990938 7.755769  7.106248  4.990938
3   8.856621 2.723064  5.504823 9.057580 7.752409 10.041043  1.974253
4   7.475424 8.910869  6.685293 2.636477 1.974253  5.450685  3.102392
5   3.832517 1.974253 10.265667 3.014929 3.014929  8.145594  4.230916
6   9.139402 1.974253 12.573464 3.594375 2.614033 12.170276  4.707005
7   7.421054 2.622193 11.256628 1.974253 2.883191 12.395667  3.502313
8   6.957665 2.721976 13.577607 3.242643 3.242643 13.746976  3.242643
9   6.727863 2.685000 13.199820 3.506601 2.685000 15.294060  3.356668
10  9.094106 2.602776  7.645452 5.957915 2.856360  2.856360  1.974253
   AC008663.1 IGHV5.10.1  VN1R53P     MUC2 HNRNPA1P57      FGB   ABCC13
1    1.974253   5.545679 1.974253 2.928090   2.654743 2.928090 1.974253
2    3.800627   7.765331 7.765331 4.475844   3.019723 5.147375 2.586544
3    1.974253   4.856782 1.974253 3.844086   1.974253 1.974253 4.059635
4    7.555178   5.271735 7.973191 2.902907   3.844644 1.974253 8.006668
5    1.974253   7.990721 1.974253 2.717920   2.717920 1.974253 1.974253
6    1.974253  11.482929 1.974253 9.538810   1.974253 1.974253 1.974253
7    2.622193   6.188444 1.974253 2.883191   1.974253 2.622193 1.974253
8    2.721976   4.678272 1.974253 6.867097   3.424637 1.974253 2.721976
9    1.974253   6.617643 1.974253 2.685000   2.685000 1.974253 1.974253
10   1.974253   6.026848 5.686548 3.923636   2.602776 5.147943 6.801112
     CYP2A6   CARTPT     GFRA1    POTEKP    SPDYC      AQP5   KLHDC7A    BPIFB1
1  3.688717 3.445442  8.717748  3.573460 3.688717 12.598583  4.288019  9.764639
2  3.529934 1.974253 13.067395  9.077315 2.586544  4.022179  8.866978  9.738349
3  4.910634 3.244417 12.972803  6.092820 5.814994  4.681422  3.021982  3.426630
4  4.363804 1.974253 15.401271 10.693337 2.636477  3.844644 12.082398  4.911387
5  3.236028 1.974253  9.461121  4.230916 3.417207 10.686839  3.014929  7.627688
6  5.220803 2.614033  7.365082  3.954451 3.362422 13.182273  2.871920 10.801366
7  2.883191 1.974253  6.661039  3.240058 8.902409 11.408501  2.883191  3.240058
8  2.721976 1.974253  7.210067  4.152451 4.797532 11.315444  6.635806  3.242643
9  2.685000 1.974253  7.227162  3.638975 5.233201  6.687514  2.685000  1.974253
10 1.974253 2.602776 12.990336  3.459548 2.602776 13.587144  9.733170  3.046656
      MAGEA3       PGR      KLK7     FOXI1 ANKRD30B LINC02747  IGLV8.61
1   8.546196  6.958749 12.309413  2.654743 1.974253  1.974253  5.053922
2   1.974253  7.544308  8.992352  4.990938 9.884772  7.604154  8.352240
3   1.974253  6.542433  2.723064  8.353819 4.800760  5.504823 10.252972
4   1.974253 15.495189  5.066784  5.797137 2.636477  9.839336  1.974253
5   1.974253  7.445251  8.965444  8.971592 4.142232  3.014929  8.439771
6   7.034072  6.155115  4.950619  8.047930 1.974253  1.974253  9.709589
7  11.148605  6.803357  8.558232  7.519442 1.974253  8.645032  8.146925
8   2.721976  5.892510 13.830795 11.963511 3.424637  3.242643 10.224186
9   1.974253  4.629385  8.150581  8.079687 1.974253  1.974253 10.705547
10  1.974253  8.682588  3.568058 12.191422 9.925281  5.051045 11.944255
        ZIC1   SORCS1 Y_RNA.699      SYT9     MUC15      PIGR LINC02188
1   6.771303 4.537480  1.974253  3.445442  9.597256  6.072397 11.816354
2   2.833902 4.814777  6.659665  3.305897  6.338918  9.801271  6.102246
3   2.723064 2.723064  1.974253  4.800760  5.637095  7.890483  1.974253
4   9.024936 9.991566  7.852129 10.306721  3.844644  5.206674  4.363804
5   8.435329 7.344573  1.974253  3.572606  6.184983  8.309909  9.294556
6   8.449117 2.871920  1.974253  2.614033  5.834721  8.327381 10.712314
7  10.365400 4.689147  2.883191  2.883191 10.168305  6.036789 11.709685
8  10.629997 3.424637  1.974253  4.241422 10.515523  8.541560 11.112398
9   8.189579 1.974253  1.974253  2.685000 11.340615 11.346558 11.750220
10  3.046656 7.356123  5.238599  2.602776  2.602776  5.267567  3.339321
      OPRPN      LRP2     KRT81      PAX7   SLC6A14 LINC01833    SPAG6
1  1.974253  9.051377 15.148280  1.974253 11.143221 10.680407 5.702875
2  7.012946  5.818601  8.882435 10.373828  6.833534  3.173241 3.800627
3  1.974253  3.426630  4.910634  1.974253  5.107413  1.974253 3.244417
4  1.974253  6.057525  6.574694  2.636477  4.481081  2.902907 2.636477
5  1.974253  7.325487 13.606466  1.974253  6.605234  3.236028 4.996580
6  1.974253 11.609449 12.747514  1.974253  9.228998  9.167264 3.362422
7  1.974253 13.398820 13.729319  2.622193 10.654389  7.681206 3.240058
8  2.721976  8.931076 16.077844  1.974253 13.152142  7.442791 3.242643
9  2.685000 13.713464  6.878816  2.685000 13.072457  8.100301 3.182196
10 1.974253  4.754083 12.080716 12.006680  7.775900  2.856360 3.568058
     SCGB2A1   PPP2R2C   GPR139     ROPN1     KRT16     ACTL8     DLK1
1   7.338266 12.204168 1.974253 11.902102 15.596776  2.928090 1.974253
2   7.770089  4.022179 1.974253  4.376094  7.179633  4.814777 3.305897
3   9.151200  9.244274 1.974253  1.974253  7.127370 10.036590 1.974253
4   6.468691 11.385932 7.309004  2.636477  6.354240  4.363804 2.636477
5   3.417207  6.455077 1.974253  9.641610 15.151551 13.276191 3.944344
6   3.594375  3.065301 1.974253  9.017872 11.583836  5.125467 6.693993
7   3.502313  9.344720 1.974253 10.626711 12.482549  7.605581 3.379129
8   5.316508  8.177429 1.974253 11.819727 11.870513  8.838864 1.974253
9   1.974253  4.144847 1.974253 10.391177 11.455548  6.368073 1.974253
10 11.013928  4.319404 1.974253  3.843955  5.866801  2.602776 3.339321
     IGHV1.3      CST1      KLK6   MS4A15   SLITRK6      TRH    IGHG1     WIF1
1   6.747556  7.918865 13.239558 4.796218  8.834927 2.928090 16.85702 5.910306
2   7.263029  7.798309  6.473583 8.226363  8.615371 4.210992 15.11403 4.426891
3  13.376662  9.789005  4.327670 1.974253  1.974253 3.426630 18.48935 3.244417
4  11.557456 10.640253  3.932899 6.661440 13.128147 4.951885 18.89088 4.166014
5   7.266665 10.668033 10.023825 1.974253  5.303358 3.014929 18.05752 5.878468
6  10.338114  7.280093  3.954451 1.974253  4.030265 1.974253 19.68822 9.050297
7   4.642212  5.478543  6.089144 4.436488  6.464021 1.974253 19.83546 8.291267
8   5.316508  8.835459 13.212115 4.056983  5.601513 3.580700 19.64769 3.841640
9   4.686374  7.099800  7.873595 5.736185  3.638975 2.969731 20.40399 1.974253
10 11.300178 10.317562  3.667250 2.856360  7.186288 2.602776 18.99584 2.856360
       CIDEC    ABCC11   CYP4Z2P     MUC16     PTPRT    INSM1   IGHV4.4
1   5.798802  4.064487  2.654743 15.369076  5.460051 2.928090  4.479301
2   7.283145  6.596702  8.623278  8.084626  7.910288 3.305897  3.529934
3   6.830471 14.122747 11.209288  6.203244  9.098042 3.021982 13.049227
4  10.158264 10.891907  9.043485  8.938386 11.932490 5.797137  8.687317
5   6.755525  5.303358  3.014929  9.980278  7.666117 8.856703  7.984652
6   5.157968  3.954451  1.974253  9.251484  3.873816 3.484371  9.895967
7   5.741740  5.052425  2.622193 11.579341  5.052425 3.240058  7.161416
8   4.959155  3.424637  3.718221 11.143017  6.798730 5.192882 10.314994
9   3.182196  3.757984  4.144847 11.928990  1.974253 3.506601  9.981721
10  2.856360 11.601821  7.740548  3.667250  5.267567 3.923636 10.347233
   LINC01016     KLK10 AP000844.2  RPSAP53       CA9     PROM1   PPP1R1B
1   1.974253 12.346643   1.974253 3.132496  5.973260 14.993798 11.263070
2   3.019723  6.575087   7.521915 8.076968  3.173241 12.830682 11.978231
3   1.974253  9.287402   8.255023 7.795237  4.617639 11.167802  7.639418
4  11.863706  4.166014   5.172983 5.505646  3.749630  6.324144 12.588511
5   1.974253 11.312830   1.974253 3.014929  8.371659 10.429835  8.290365
6   1.974253  7.250611   1.974253 7.508962  5.752606 15.008453 13.270926
7   2.883191  8.791313   1.974253 7.652587  6.565902 14.536573 12.592063
8   3.242643 11.490162   1.974253 4.056983 11.067394 14.429715 11.609579
9   1.974253  4.442610   1.974253 8.090031  9.558812 13.533722 10.069262
10  2.602776  2.602776   6.567181 5.377849  7.495970  7.093220 11.256615
     SLC7A4      ORM1     LY6D     NLRP2    MKRN3    MAGEA6 AL021395.1
1  5.821809  1.974253 5.202663  6.110208 9.106026  7.248121   2.654743
2  3.019723 11.251237 3.423612 13.218577 7.249461  1.974253   1.974253
3  2.723064  6.224344 1.974253  8.176214 1.974253  1.974253   3.426630
4  7.354305  3.102392 2.636477 12.020843 3.533619  1.974253   5.962898
5  2.717920  3.417207 7.941445  7.803464 3.236028  2.717920   3.944344
6  1.974253  3.224804 9.847589  7.941295 8.313198  6.122856   1.974253
7  2.883191  3.714841 8.921720 11.667413 8.676636 10.536429   2.622193
8  3.242643  8.821758 8.094391 11.986333 9.421297  2.721976   3.580700
9  6.086687  3.966108 8.232228 12.242029 8.095175  1.974253   1.974253
10 2.602776  6.140032 2.602776 10.172616 2.856360  1.974253   1.974253
       RIMS4 ARHGAP36   SDR16C5     KRT17     MSMB      GRPR    WDR72
1   6.072397 1.974253  9.529334 15.756593 3.445442  3.132496 7.346188
2   6.246142 1.974253  5.762137 12.396737 8.073124  4.210992 2.586544
3   6.092820 3.244417 11.024465  5.504823 2.723064  5.896268 3.244417
4  12.822128 3.102392  3.266704 10.640253 9.160687 11.285303 2.636477
5   3.417207 1.974253  6.967458 14.651237 1.974253  3.014929 2.717920
6   4.234206 1.974253  1.974253 12.531206 3.484371  3.362422 3.224804
7   3.714841 1.974253  4.642212 10.318820 5.674644  3.976682 2.622193
8   7.673946 1.974253  3.841640 14.478073 8.776312  4.152451 5.994122
9   5.736185 1.974253  1.974253 13.010008 5.115192  3.866385 1.974253
10  9.890003 5.351073  5.903946 12.088266 5.455268  4.910458 2.856360
   AC100801.1 LINC01488 MRPS30.DT     ZIC2       LBP        CP BMPR1B.DT
1    1.974253  4.418512  8.664912 7.719887  3.793839 14.273311  1.974253
2    9.133762  5.800026 11.925382 8.201928  6.175998 13.053076  1.974253
3    1.974253  1.974253  5.279990 6.897576 10.864463 10.901675  1.974253
4    4.015388  8.319873  7.222064 4.481081  8.111491  9.179676  1.974253
5    1.974253  3.709572  6.917327 7.916161  5.263413 10.111346  1.974253
6    3.484371  4.234206  5.157968 2.871920  3.954451 16.002751  2.614033
7    1.974253  6.071904  5.340639 3.078798  3.502313 13.604183  2.883191
8    2.721976  3.020490  6.065861 3.424637 13.519786 13.512876  1.974253
9    1.974253  1.974253  4.058570 1.974253  9.316121 10.265651  1.974253
10   3.998580  2.602776  6.301601 4.910458  6.688897  8.563397  6.868220
    PGLYRP2  IGLV3.25  IGHV1.18  RHOXF1P3 RN7SL314P    LRP1B    NPY5R  IGLV3.10
1  3.132496  8.040990  6.381406  3.445442  1.974253 4.931011 2.654743  4.748212
2  4.210992  8.695065  8.758669  6.102246  6.008422 4.737546 4.323300  6.190304
3  3.244417 11.149356 12.473105  3.021982  1.974253 2.723064 2.723064 11.382489
4  8.375840 10.936004 11.699815 10.190868  6.509363 4.235276 6.339271 10.223768
5  4.392248 13.938792 13.045948  5.416842  1.974253 4.392248 4.047086  9.856179
6  3.594375 10.148464 14.459380  2.871920  1.974253 1.974253 4.661883 11.102316
7  3.240058 13.474425 12.327625  3.078798  1.974253 4.543228 4.490907 12.158371
8  4.477391 13.397124 11.335467  4.152451  1.974253 4.403249 8.401212  6.853681
9  2.969731 14.038481 12.443988  2.969731  1.974253 2.685000 7.553049 17.171862
10 7.155929 11.684178 13.000109  3.923636  3.203721 4.529855 1.974253 10.257216
        ROCR  IGLV6.57     GPR26     PEG10   MAB21L4       PI3     FABP4
1  11.458338  9.185407  9.885017 14.235846  7.239640  5.775418  8.464892
2   4.923121  9.993496  3.019723  8.297280  7.504888  5.288149  9.485685
3   1.974253 11.118079  1.974253 15.426247  9.170302  3.021982 10.152269
4   4.423717 10.061483  3.102392 11.959599 10.893824  3.408295 13.385142
5   8.638344  9.224206  5.824891 10.741828  8.487752  6.472557 10.134820
6   8.986937 10.389706  4.987378  7.795854  8.262429  6.715749  7.126254
7   9.884449 10.950736  3.240058 13.974106  3.502313  6.844459  5.719723
8   8.428589 12.297841  7.261748 14.949015  9.396096 11.174347  7.681553
9   6.107264 16.954459 10.356260 14.239838  7.977537  6.673808  4.058570
10  1.974253 14.423961  6.967734 11.219967 15.238253  2.856360  5.789483
        MSLN   IGLV3.1   IGHV1.2   ABCC12    PHGR1     CST4     DSG1     CDH19
1  15.180777  8.756885  5.130258 2.654743 2.654743 3.300671 4.887515 10.663080
2   3.627195  8.986233  7.681954 3.800627 4.210992 3.173241 4.990938  3.529934
3   5.238785 14.167058  9.724010 8.344238 5.060748 4.681422 1.974253  1.974253
4   4.301089 13.124068 13.554634 4.826679 6.720347 5.172983 5.103082  3.646549
5  14.937450 13.023746  8.876498 1.974253 1.974253 4.314019 5.878468  8.140145
6   4.169698 13.124567 12.124313 1.974253 1.974253 3.065301 4.295713  7.093412
7   4.436488 15.451922 13.582261 2.622193 3.078798 3.379129 6.477158  5.555282
8  11.874665 14.148950 11.541735 1.974253 1.974253 4.056983 6.088998  5.393302
9   7.227162 13.986928 14.895181 5.270441 2.685000 2.969731 4.686374  1.974253
10  3.046656 12.572447 10.597013 9.022345 1.974253 4.754083 1.974253  1.974253
       FOLR1      FGG   HORMAD1      LEP    NBPF4     IL20    CXCL17  IGHV3.48
1   8.117015 2.654743  2.928090 4.288019 2.654743 3.445442  3.445442  8.211077
2   5.781207 4.210992  3.717027 5.487770 5.925247 5.117451 13.480380  6.542042
3   1.974253 3.244417  3.582869 5.698892 1.974253 3.426630 10.020140 10.237531
4   6.776936 1.974253  3.646549 9.054957 3.533619 4.911387  7.782046  9.689879
5   4.785476 1.974253 10.479215 6.489827 6.955088 2.717920  3.709572  8.483455
6   5.189736 1.974253 13.178733 3.787612 2.871920 3.362422  3.065301 13.011647
7   9.025777 1.974253 10.129386 3.379129 5.052425 3.379129  4.379783 11.694329
8  11.040294 1.974253 10.863538 5.969387 1.974253 3.953879  7.899439 11.892386
9   7.983098 3.866385  4.302034 5.736185 6.631891 3.638975  9.448604 15.612777
10  2.856360 5.267567  4.136486 1.974253 8.010215 4.794859  6.720379 13.145555
    IGHV3.21  SCGB3A1    ABCC8      LCN2      EDN3    PTPRZ1    EEF1A2
1   9.538002 3.688717 5.092608 13.265101  2.654743 14.078772 12.976386
2   7.973526 3.717027 6.508220 10.360488  5.890575  6.132204 11.604049
3  11.518063 3.956435 5.012477  9.427807  1.974253  2.723064  8.377495
4  10.841128 5.962898 9.951616  6.093692  4.235276  4.301089 11.592122
5   9.452355 5.555282 1.974253  8.161819 10.038542  6.878550  8.778202
6  13.304580 6.481788 2.614033  9.242158  2.871920  3.594375  4.874064
7  12.733828 3.379129 2.622193 12.531291  3.502313  9.511722  8.215042
8  13.385822 8.000207 2.721976 15.334897  3.020490  7.571212  8.291173
9  13.873367 1.974253 2.685000  9.018772  8.799988  3.182196  3.182196
10 13.482641 3.843955 8.064315  4.200317  1.974253  3.046656  4.910458
   IGHV1.69D      DSC3   CYP2A7    HOTAIR     MUC19  PLA2G2A     KRT4     SIM2
1   6.033562 10.975978 3.300671  5.337199  3.132496 5.910306 5.727468 9.483406
2   4.737546 11.313648 2.833902  8.810381 10.001607 7.302984 1.974253 9.429672
3   9.842718  3.844086 3.426630  8.819075 10.608489 8.726241 8.176214 7.464692
4   9.646555  8.390404 1.974253 10.614899  2.636477 8.554592 1.974253 4.689049
5   9.066559 10.801716 1.974253  6.682341  1.974253 8.101414 4.895023 9.352241
6  13.970942 12.506880 2.614033  6.999033  1.974253 5.280959 3.594375 9.175127
7  11.324849 12.804447 2.883191  6.717455  1.974253 4.436488 5.281486 6.661039
8  11.002844  6.970150 1.974253  7.963239  2.721976 4.403249 6.712156 9.676649
9  11.796736 10.273612 1.974253  4.058570  1.974253 1.974253 2.685000 8.218151
10  9.979340  5.598493 1.974253  8.614494  4.428728 6.140032 2.602776 1.974253
       CLIC6     IGHM AC078993.1     TNNT1 LINC02562    CIDEA  IGHV5.51
1   8.331645 11.68400   2.928090  7.361903  1.974253 4.288019 12.198986
2  10.994052 11.87770   3.305897  9.289291  1.974253 5.703347  8.280709
3   7.550540 15.15691   4.155189  5.012477  1.974253 5.060748 12.218083
4   8.464546 18.77424  10.036239 10.697735  1.974253 8.935906 12.295902
5   6.605234 14.24399   1.974253 11.159707  8.224952 6.163598 12.710912
6   7.365082 16.55197   2.614033  8.946163  1.974253 3.594375 14.687581
7   7.217222 13.16589   3.078798 11.402500  6.437380 4.193638 14.803136
8   7.704138 16.01405   3.242643  4.678272  2.721976 1.974253 12.067659
9   6.754147 15.76292   2.685000  5.506836  1.974253 1.974253 15.102249
10  7.186288 16.46591   4.577731  5.238599  7.117057 1.974253 13.881899
     BRINP2       ER
1  5.014139 Negative
2  7.956932 Negative
3  1.974253 Negative
4  3.266704 Negative
5  4.466175 Negative
6  1.974253 Negative
7  4.258627 Negative
8  6.281966 Negative
9  5.506836 Negative
10 6.092615 Negative</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>er_folds<span class="sc">$</span>splits[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> <span class="fu">testing</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            ID   CLEC3A   SCGB2A2      CPB1     GSTM1     TFF1    S100A7
1  Training_11 2.925978  4.060384  3.570125  2.925978 3.442332 11.454923
2  Training_19 1.974253  9.958888  3.742617  5.993970 3.375647  6.890274
3  Training_20 1.974253  5.108298  2.799514  6.339706 2.561722  2.561722
4  Training_21 2.875813 10.638139  5.942479 12.452599 2.875813  4.363248
5  Training_22 1.974253  1.974253  8.276546  1.974253 2.945428  3.600794
6  Training_61 1.974253  3.557730  3.961746 11.389608 1.974253  7.861741
7  Training_68 3.391029 12.108652  8.519248  1.974253 6.520717  3.544079
8  Training_74 7.401455 13.309164 11.950340  8.095732 5.902822  7.095775
9  Training_77 4.371477  3.311226  3.456977  6.150631 3.701796  2.660432
10 Training_85 2.788086 10.734562  3.711402  7.437853 1.974253 10.128852
     SCGB1D2    KCNJ3 LINC00993   VSTM2A ANKRD30A DSCAM.AS1 AC093001.1
1   2.925978 3.570125  5.620797 1.974253 5.697180  1.974253   3.297826
2   7.058938 2.804049  3.742617 2.804049 3.132637  1.974253   3.375647
3   3.471244 4.760553  1.974253 2.561722 1.974253  1.974253   1.974253
4   7.605945 3.069964  1.974253 1.974253 1.974253  1.974253   2.616851
5   2.945428 3.717628  1.974253 1.974253 1.974253  1.974253   3.324013
6   3.557730 1.974253  2.918135 3.776393 1.974253  1.974253   5.945629
7  10.193870 3.212737  6.488049 1.974253 6.090926  1.974253   6.068971
8  10.359322 7.079467  6.044392 4.149067 6.173221  4.010580   5.902822
9   4.992253 2.660432  3.141857 2.660432 2.660432  1.974253   2.660432
10  6.627260 3.544867  1.974253 1.974253 2.788086  1.974253   2.788086
         PIP SERPINA6     MUCL1     PRAME   PSPHP1  CYP2B7P     MUC6  UGT2B11
1   2.653211 3.570125  7.207607 12.675465 8.907780 3.129973 4.743279 3.129973
2   5.400261 1.974253  5.763899 12.755910 9.397795 3.573645 3.132637 2.804049
3   1.974253 1.974253  4.294449  3.881454 8.392830 4.760553 4.960861 1.974253
4   5.168309 1.974253  1.974253  1.974253 8.780521 3.794802 4.419701 1.974253
5  11.157188 2.667333  5.173422 11.669805 1.974253 5.349027 3.324013 2.945428
6   1.974253 4.267227  1.974253 14.010430 2.918135 8.491247 3.776393 1.974253
7  10.747475 4.275978 11.481805  9.380460 1.974253 5.176254 7.459658 6.046676
8   5.975356 3.211697 11.257896 11.993917 9.502314 5.367397 5.066653 5.066653
9   4.435423 1.974253  4.861224 13.317339 2.660432 3.807544 3.701796 1.974253
10 12.483028 2.788086  7.736086 12.928707 8.142359 6.242202 3.857646 2.788086
       CST9   HMGCS2  SLC30A8    TFAP2B    CASP14    HOXB13    CYP4Z1
1  1.974253 2.925978 6.159180  5.769680  4.350365  1.974253  3.442332
2  1.974253 3.742617 3.573645 11.998716 10.351546  8.674739  1.974253
3  1.974253 2.799514 1.974253  3.949206  1.974253  7.072262  6.646526
4  1.974253 7.600062 4.883915  6.229136  6.650033  1.974253  3.069964
5  1.974253 4.455855 4.455855  8.595543  7.050255  4.632252  3.470948
6  3.776393 5.986014 9.929351  6.337401 11.943787  9.311113  5.279107
7  2.703658 9.235745 3.800333  8.440477  3.679075  6.962245  7.917149
8  5.132155 4.768857 4.925687  9.725540  6.664114  8.508638  4.591945
9  1.974253 3.995293 4.906282  7.461397  3.904923 10.620116  2.935926
10 1.974253 7.530971 3.544867  7.795671  5.103091  3.349906 11.484036
   AL157387.1      GP2 AC093838.1    BRINP3     FDCSP    CALML5    ADIPOQ
1    3.886815 1.974253   2.653211 10.571638 13.795640  9.380367 11.680900
2    1.974253 1.974253   1.974253 14.543043  3.132637  6.562893  4.253543
3    1.974253 1.974253   1.974253  6.024398  5.785029 13.370636  3.254542
4    1.974253 1.974253   1.974253  1.974253  2.616851 11.119790 10.877558
5    1.974253 1.974253   4.517228  5.444800  5.444800 10.279445  4.928934
6    1.974253 1.974253   5.945629  3.287253  5.343058 11.391481 13.670470
7    4.275978 5.539559   1.974253  4.353367 12.587153  8.431930 11.682022
8    4.494116 3.677710   4.768857 12.295808  8.789750  6.426682  6.685835
9    2.660432 4.610929   2.660432  4.435423  8.780917 13.226734  8.230566
10   1.974253 3.349906   1.974253 11.017013 19.253927 12.794623  7.968356
      SLC5A8    GRIA2    KLK11     ADH1B    KCNC2     DHRS2     FABP7   CYP4F8
1   7.003302 7.233350 6.263023 11.486911 1.974253  4.413946 12.318906 1.974253
2   1.974253 1.974253 5.202960  4.537741 1.974253  2.804049  9.607539 3.132637
3   5.917653 5.559669 6.024398  2.799514 2.561722  2.978430 13.518784 1.974253
4   5.885350 6.650033 8.940347  3.230074 2.616851  5.068298  3.368196 1.974253
5   3.717628 9.202824 5.015255  5.935980 2.945428 12.315879  7.444299 3.324013
6   2.918135 6.833272 8.674724 13.629792 1.974253 11.262330  2.918135 4.515259
7   5.802239 4.685131 9.824289 12.261772 3.800333  5.295355 12.145382 4.902721
8  10.284887 6.769583 3.855261  6.142083 1.974253  8.556521 11.479758 2.862244
9   3.585824 8.281858 4.949928  7.809050 3.141857  3.141857 11.252538 1.974253
10  5.483665 4.865648 8.168350  9.145447 2.788086  3.857646  9.474499 3.711402
       SOX10 AC093297.1     KRT14 LINC02224   CEACAM5 AC061961.1      CGA
1  14.496237   1.974253 11.816495  1.974253  5.745921   1.974253 3.129973
2  13.205717   1.974253 13.481418  1.974253  3.132637   1.974253 2.804049
3  13.882646   2.799514 10.707224  2.561722  3.254542   1.974253 1.974253
4  13.692496   1.974253 15.728283  2.616851  8.490620   1.974253 1.974253
5  13.597464   1.974253  8.396187  1.974253  3.153202   1.974253 4.928934
6   1.974253   3.287253 12.579642  1.974253 10.722459   1.974253 4.818868
7  12.924886   2.703658 16.346511  2.703658  3.800333   1.974253 1.974253
8  14.316448   1.974253  9.647550  1.974253  4.683241   4.768857 4.010580
9  14.190244   2.935926 12.293027  1.974253  2.935926   1.974253 2.935926
10 13.656415   1.974253 10.263028  1.974253  6.086883   1.974253 3.349906
      NXPH1     TFF3     SYT13 AC044784.1 LINC01087       DCD    GLRA3    OBP2B
1  5.124971 4.413946 10.558978   1.974253  2.653211  1.974253 3.685189 4.925900
2  1.974253 4.620609  7.381660   3.573645  1.974253  3.742617 1.974253 7.625771
3  2.561722 3.126459  3.881454   1.974253  1.974253  2.978430 2.561722 4.686792
4  1.974253 4.242654  4.038167   1.974253  1.974253  3.600947 3.069964 5.698833
5  3.717628 7.603699 10.422355   1.974253  4.391566  2.667333 3.600794 6.213501
6  1.974253 6.172443  7.828934   1.974253  7.828934  2.918135 5.861269 1.974253
7  1.974253 7.232313  6.274711   2.703658  1.974253  3.212737 7.390734 7.451221
8  4.494116 7.249438  8.877192   6.044392  3.855261  6.596908 8.862981 7.896167
9  1.974253 3.456977  7.438969   1.974253  1.974253  4.079688 5.843121 4.496477
10 1.974253 7.047407  8.193881   1.974253  1.974253 15.317820 3.544867 7.338300
        TAT    MYBPC1 LINC01291 SLC25A24P1     BEX1     KRT6B    COL2A1
1  4.413946  9.489362 10.700188   6.616438 9.176661 16.255147  4.060384
2  4.537741  1.974253  7.606111   2.804049 1.974253 14.997411  4.449504
3  3.881454  6.408785  8.159405   3.733763 2.799514  9.039967 12.157290
4  6.890940  7.054846  5.979336   3.069964 4.843924 11.908848 11.729897
5  4.013141  3.153202  3.600794   1.974253 3.324013 12.416400  8.196963
6  2.918135  4.123427  4.906864   4.267227 5.675956 14.561921  6.100755
7  8.570767 11.106857  8.515207   2.703658 3.800333 13.869232  5.507142
8  4.683241  7.189925  5.902822   5.066653 4.388685 10.028975  8.234674
9  3.904923  6.863141  2.660432   2.660432 4.371477  7.989600 12.104715
10 4.215863  7.599454  1.974253   5.441324 4.799386 10.166407  5.047398
       STAC2    CRABP1   CRISP3     OLFM4    SRARP ARHGAP40    PVALB   SIRLNT
1   7.790639 15.035192 3.570125  9.440502 1.974253 5.860954 3.129973 3.129973
2  17.774314 14.551256 5.573426 12.525221 2.804049 5.899861 2.804049 2.804049
3  13.830349  6.517011 8.630227  2.978430 2.561722 2.978430 2.799514 1.974253
4   4.242654 11.150796 3.230074 10.921843 1.974253 7.502507 1.974253 2.616851
5  11.982495 12.532150 2.945428  9.686664 2.667333 2.945428 6.844934 6.591586
6   3.557730  3.287253 6.368233 17.875419 1.974253 7.985977 5.141801 4.515259
7  13.349095  5.176254 2.995361  9.520058 5.176254 7.492922 4.193801 1.974253
8  16.282282 15.944458 3.468942  7.189925 4.494116 6.619661 5.254668 6.573790
9  13.284713 15.064450 2.935926  3.995293 3.311226 5.149820 2.660432 1.974253
10  9.355129 10.868394 9.886218 11.559616 1.974253 4.928896 3.711402 1.974253
       MUC5B    GABRP     NPY1R    CBLN2      KRT5     CLCA2  S100A7A    HOXC12
1  12.938992 16.01539 11.712873 4.139098 11.670737  4.213335 6.921183 10.255337
2  10.914773 15.69982 14.507318 3.375647  8.536208  8.131192 5.763899  6.165447
3  12.142400 15.34197 12.605437 4.991616 10.524550  3.471244 1.974253  8.978945
4   9.050113 13.57074  4.802748 3.230074 13.148819  4.242654 1.974253  5.653853
5   8.183262 15.01977  4.177786 2.945428 15.489433  2.945428 1.974253 10.438782
6   2.918135 13.00364  5.462902 1.974253 18.170165 13.433331 1.974253  9.574488
7  11.914505 15.09909 11.269183 3.679075 15.458530  7.043105 1.974253  8.921533
8  13.503552 13.92430 12.669927 2.862244  7.427451  4.010580 3.677710  9.189589
9  13.806340 15.71535  6.851859 5.073246 14.841318  6.696535 2.660432  9.293162
10  8.660111 15.94072 10.935188 1.974253  8.748903  8.545801 4.579104  8.795505
         LTF     KRT6A    BMPR1B SERPINA11      KLK5    PYDC1      TCN1
1  11.458563 13.983817  8.831225  6.246229 12.563305 4.642058 11.295656
2   8.525785  8.335987  7.014863  1.974253 13.455055 3.132637 10.228409
3   6.774080  4.864310  7.624729  2.978430  7.319172 1.974253  4.960861
4   8.880952  4.242654  4.843924  3.230074 14.061779 2.875813 11.893303
5  10.918300  8.297895  8.280841  3.153202  9.142746 5.210334 11.838070
6   6.694324 16.846999  5.861269  3.287253  7.201366 4.515259  8.585901
7  13.874071  9.080357  9.295855  5.632603 13.563427 3.212737  6.825224
8   8.738778  5.254668  8.759384  3.468942 12.643154 3.211697 10.564963
9   8.810325  5.723935 10.811135  3.904923  8.881324 5.291510  1.974253
10 15.892419  8.594536  5.397681  3.110882 10.615648 7.437853  7.900192
      TRARG1 C10orf82 IGHV7.4.1 AC008268.1     KIF1A     CST5    ASCL1
1   7.416773 2.925978  2.925978   1.974253 12.024489 1.974253 1.974253
2   2.804049 1.974253  3.132637   1.974253 10.805926 1.974253 1.974253
3   1.974253 2.799514  3.652606   1.974253  8.451184 1.974253 3.126459
4   1.974253 2.616851  3.794802   1.974253  5.261687 1.974253 1.974253
5   3.824130 3.153202  6.000141   3.324013 13.400147 1.974253 5.563206
6  11.330304 6.206986  3.287253   2.918135  8.389943 1.974253 1.974253
7  10.121308 4.106144 12.881608   5.473967  4.495925 2.995361 3.800333
8   4.388685 3.677710  1.974253   5.826411 11.502852 1.974253 4.388685
9   6.605390 4.079688  4.233623   2.660432 12.194991 1.974253 5.420284
10  5.816887 1.974253  6.086883   3.110882  5.641445 1.974253 2.788086
        DSG3   BPIFB2 IGHV1.69.2   CEACAM6   OAZ1P1   SLC34A2   LRRC31   CDC20B
1  10.574995 3.297826   1.974253  5.721759 3.129973  8.782682 4.837589 3.570125
2  10.715962 4.620609   1.974253  9.135243 1.974253 10.326244 3.573645 3.742617
3   4.013519 2.561722   9.943741  5.136014 2.978430  8.836105 3.733763 3.254542
4  12.258115 1.974253   8.282358 13.379284 1.974253  7.696953 2.616851 2.616851
5   5.173422 2.945428   1.974253  6.809661 1.974253 11.667347 3.824130 2.945428
6  11.074575 4.267227   1.974253 12.791995 1.974253  4.123427 4.267227 5.279107
7  13.814601 5.332925   4.275978  6.331089 3.910693 12.622086 3.679075 2.995361
8   8.484087 4.683241   1.974253  6.374097 2.862244  7.686168 3.468942 1.974253
9   8.010024 1.974253   1.974253  2.935926 1.974253 10.123571 1.974253 2.935926
10  5.156661 1.974253   6.836474 10.409125 1.974253 12.880397 5.156661 2.788086
        ELF5 LINC00052 AC008663.1 IGHV5.10.1  VN1R53P     MUC2 HNRNPA1P57
1  13.908640  7.755550   2.925978   3.297826 3.570125 3.976556   1.974253
2  12.768949  3.890877   1.974253   5.305048 1.974253 3.573645   2.804049
3  13.627300  1.974253   1.974253  11.480580 1.974253 2.978430   3.733763
4  12.365365  1.974253   1.974253   9.269776 2.616851 1.974253   2.616851
5  14.027793  3.153202   1.974253   4.928934 1.974253 1.974253   2.667333
6   8.484242  9.070449   2.918135   1.974253 1.974253 6.644857   1.974253
7  11.858207  1.974253   4.495925   5.775330 3.212737 2.995361   2.703658
8  13.054278  4.494116   4.010580   7.514912 1.974253 3.855261   2.862244
9  14.056752  3.807544   1.974253  10.615126 1.974253 3.141857   2.660432
10 13.821213  1.974253   1.974253   7.778897 1.974253 2.788086   1.974253
        FGB   ABCC13   CYP2A6   CARTPT     GFRA1   POTEKP    SPDYC      AQP5
1  1.974253 3.297826 3.129973 1.974253 11.388261 3.442332 1.974253  9.471458
2  2.804049 4.143545 4.023423 1.974253  4.973479 3.890877 5.149006  6.661550
3  1.974253 2.561722 4.724165 1.974253 13.032395 3.126459 4.724165 14.112008
4  1.974253 2.616851 3.701783 3.701783  6.727713 2.616851 8.425594 10.802082
5  1.974253 3.600794 4.928934 2.667333  8.959320 4.883687 5.349027  9.057705
6  3.557730 6.980053 6.100755 2.918135  8.896704 7.316326 7.749351  6.240718
7  2.995361 2.703658 5.046147 1.974253 11.180769 4.353367 5.632603 10.791566
8  3.677710 5.132155 4.388685 1.974253  8.327300 4.997944 3.211697 11.834961
9  1.974253 1.974253 3.141857 1.974253 11.870389 3.141857 2.660432 11.180050
10 1.974253 4.409699 5.156661 1.974253  8.748903 3.544867 7.033324 12.922279
    KLHDC7A   BPIFB1    MAGEA3       PGR      KLK7     FOXI1 ANKRD30B LINC02747
1  1.974253 3.297826 11.416607  6.246229 13.587208  7.572760 3.570125  2.653211
2  6.680494 4.842896  1.974253  5.445571 10.814514  6.192133 1.974253  1.974253
3  1.974253 2.799514  1.974253  5.599792  7.754011 10.931661 1.974253  2.561722
4  1.974253 6.182537  1.974253  6.694934 14.260637  5.200127 2.875813  2.616851
5  6.213501 1.974253 12.201328  4.098072  7.019719  7.582973 4.632252  1.974253
6  6.206986 6.485345 10.553743  3.776393 10.865431  7.408052 2.918135  2.918135
7  5.295355 4.999953  2.703658 10.191340 10.621603 11.797130 2.703658  4.353367
8  5.471813 6.077701  8.819490  6.400630 12.414228  6.642059 4.591945  5.194742
9  4.435423 2.935926  2.660432  5.698855  8.433200  8.506902 1.974253  1.974253
10 4.107143 3.544867  1.974253  6.002419  6.510010  9.502859 1.974253  1.974253
    IGLV8.61      ZIC1   SORCS1 Y_RNA.699     SYT9     MUC15      PIGR
1   8.689725  9.429416 2.925978  1.974253 2.925978  6.493865  8.014912
2   5.092901  9.995227 1.974253  1.974253 3.132637 10.321733 10.233221
3  11.148266  9.890515 3.254542  1.974253 2.561722  6.946431  3.254542
4   8.558917  9.293658 6.638585  1.974253 2.875813  9.883553 10.993133
5   4.517228  9.149807 8.988496  1.974253 3.470948 10.616641  7.307791
6   3.776393 10.698451 3.557730  1.974253 4.267227  7.665117  7.149144
7  13.345118  9.220938 7.699369  3.212737 5.090868  9.389316 11.037429
8   6.233539  3.468942 3.468942  1.974253 3.677710 12.802720 11.799961
9   6.522387  9.746923 7.809050  1.974253 4.496477 10.733442  4.496477
10  8.443095  5.258053 4.215863  1.974253 3.110882  9.209423 11.697615
   LINC02188     OPRPN      LRP2     KRT81     PAX7   SLC6A14 LINC01833
1  11.972611  5.594397  7.688915 14.098050 1.974253 11.579244  6.717259
2  10.177718  5.149006  8.774958 11.226908 3.573645 10.901734  4.842896
3   6.907227 10.050044  3.809820  4.648376 2.561722 10.538714  6.984595
4  10.709522  2.875813 14.216094 14.930818 1.974253  9.525426 10.117677
5  12.091938 11.940020 10.138662  6.505978 2.667333 10.604701  9.468704
6   2.918135  1.974253  6.593627  6.876758 9.774334  3.287253  3.961746
7   8.503018  2.995361  8.722153  6.901574 1.974253 10.502178  8.143435
8  10.217612  5.975356  8.769577 14.677023 3.855261 10.585301  7.205035
9  11.059714  2.660432 13.334392 13.682008 2.660432  7.602611  6.829027
10 11.925644  3.110882  9.050827  6.682484 1.974253 12.941008 10.585829
      SPAG6   SCGB2A1   PPP2R2C   GPR139     ROPN1     KRT16     ACTL8
1  3.442332  3.570125 10.471612 1.974253  9.283133 11.992833  9.777146
2  2.804049  6.969396 10.624904 1.974253 11.296766 11.594140 10.445774
3  3.368340  3.368340  5.108298 1.974253 10.699794 10.699794  6.150897
4  3.490572  8.245153  3.230074 1.974253  8.856492 11.405096  5.261687
5  3.324013  1.974253  6.430525 1.974253  8.314749  6.977972  8.776756
6  8.189192  7.535545 10.310146 1.974253  1.974253 16.703649 10.647636
7  4.275978  8.159057  8.823652 1.974253 10.640271 15.131817  1.974253
8  3.211697 10.239776 11.041541 1.974253  9.680335 10.605357  7.095775
9  3.904923  6.187064 11.561375 2.935926 12.309596 12.689611  6.874334
10 3.349906  8.355316  4.928896 1.974253 11.259174 11.609075 10.015317
        DLK1   IGHV1.3      CST1      KLK6   MS4A15   SLITRK6      TRH
1   9.418244  4.588507  4.283615 10.963626 4.882444  9.071242 2.653211
2   1.974253  7.044396  8.960527 12.203037 5.092901  6.138255 1.974253
3   1.974253  9.021747  3.254542  5.409590 4.133227  8.679876 2.561722
4   1.974253 10.678464  5.534769 10.984130 3.701783  9.383706 2.616851
5  11.746720  5.822203  7.589915  8.404084 8.073655  9.643961 3.717628
6   4.624089  3.961746 10.644498  9.425225 8.686978 10.157812 4.396952
7   4.851438 13.180936  9.049961 10.917817 3.212737 12.563629 4.561947
8   6.707234  5.745682  5.615321 10.009756 4.274274 13.234326 6.110253
9   2.935926 10.721883  8.221838  7.731216 2.660432  5.420284 3.311226
10  3.110882  9.432226  9.264673  9.543140 3.349906  6.086883 4.989401
       IGHG1      WIF1     CIDEC    ABCC11  CYP4Z2P     MUC16     PTPRT
1  14.801301  5.232143  7.657577  4.532774 3.685189 10.634129  6.973053
2   9.303323 11.014117  5.034461  8.792455 1.974253 10.998185 11.221268
3  17.381017  8.802346  4.242865  3.949206 2.799514  8.854753  3.652606
4  15.696210  7.916397  4.473885 13.906190 2.616851  9.429196  9.686729
5  12.344882  4.013141  4.686335  4.098072 2.667333  4.836919  5.774029
6   9.311113  3.961746 12.217366  5.519231 4.396952  4.123427  2.918135
7  19.665672  7.549349 10.241109  6.974078 5.217082 13.466331  7.451221
8  14.475505 11.794356  5.312155  6.477410 3.468942  9.921096  7.205035
9  17.150515  8.786847  6.591885  6.055281 2.660432 12.552339  9.166991
10 18.291073  6.627260  7.231355  4.989401 7.129134  5.397681  5.397681
      INSM1   IGHV4.4 LINC01016     KLK10 AP000844.2  RPSAP53       CA9
1  4.139098  5.124971  2.653211 12.000969   1.974253 1.974253  8.923669
2  4.698751  5.092901  2.804049  9.805778   1.974253 6.192133 10.223581
3  3.652606 10.501726  1.974253  6.096049   1.974253 4.074757  5.316118
4  1.974253  8.271297  1.974253  9.181949   1.974253 6.831878  3.881267
5  3.324013  6.139538  2.945428  5.672511   1.974253 9.962731  5.135517
6  3.287253  3.961746  1.974253  9.187193   4.123427 1.974253 10.007151
7  5.176254 10.705545  5.134209  9.705219   1.974253 7.484678  5.217082
8  3.468942  6.502117  3.677710  9.427040   4.010580 1.974253  7.728597
9  4.554903 10.072302  2.660432  8.844841   1.974253 4.435423  9.328145
10 5.156661  8.660111  3.544867  5.973120   2.788086 2.788086  9.826272
      PROM1   PPP1R1B   SLC7A4     ORM1      LY6D     NLRP2    MKRN3    MAGEA6
1  12.99724 12.376559 3.129973 1.974253  3.570125 11.821804 9.456973  2.925978
2  13.18227 18.251581 1.974253 8.317976  8.434006 12.571874 8.739315  2.804049
3  12.00215 13.219871 4.864310 1.974253  1.974253  6.842193 9.159782  1.974253
4  12.56574 10.242789 3.368196 5.404792  3.230074 11.758889 3.069964  1.974253
5  10.28800  9.256144 2.667333 1.974253  5.724173 10.646068 9.705955 11.807142
6   3.55773  9.697684 2.918135 3.557730 14.980916 12.999036 8.171857  2.918135
7  11.77365 11.172456 7.031828 4.426528  5.602259  8.890678 7.345920  1.974253
8  11.83920 17.641447 1.974253 8.484087  3.677710 13.860615 9.391181  7.728597
9  13.26015 11.523259 4.079688 3.701796  5.291510 11.606075 2.935926  2.660432
10 14.16755 14.609089 3.349906 6.975570  9.323677 10.076636 5.483665  1.974253
   AL021395.1    RIMS4 ARHGAP36  SDR16C5     KRT17     MSMB     GRPR     WDR72
1    4.283615 8.451121 5.161608 8.255737 13.649340 5.594397 3.442332  1.974253
2    7.514162 5.652694 1.974253 1.974253  9.428892 5.092901 3.132637  4.253543
3    1.974253 3.733763 2.561722 4.568160 11.456366 2.978430 3.809820  1.974253
4    5.291489 4.419701 3.230074 1.974253 12.517218 4.843924 4.525985  2.616851
5    3.153202 5.891549 1.974253 4.972761  8.885098 7.389647 4.177786 10.942898
6    2.918135 5.724588 1.974253 5.904071 19.029989 7.904353 1.974253  9.944740
7    3.544079 5.046147 4.952183 9.183242 15.823383 3.212737 4.952183  4.685131
8    2.862244 6.077701 4.149067 4.010580  9.884048 4.683241 5.615321  3.468942
9    4.716551 7.385239 5.420284 5.291510 13.092698 5.222443 3.701796  4.610929
10   1.974253 3.857646 1.974253 6.059278 11.423959 9.515568 3.711402  2.788086
   AC100801.1 LINC01488 MRPS30.DT      ZIC2       LBP        CP BMPR1B.DT
1    4.139098  2.925978  6.359855  3.297826  9.049684  7.032928  2.925978
2    3.375647  1.974253  5.400261  2.804049  7.298339  8.979746  1.974253
3    2.561722  1.974253  9.403013  1.974253  3.254542  5.869349  2.561722
4    3.069964  2.616851  6.032901  8.241379  4.624566 10.691659  1.974253
5    2.945428  2.667333  5.845694  1.974253  3.717628  4.177786  1.974253
6    3.557730  1.974253  5.771617  9.882174 10.335664  9.768580  1.974253
7    3.800333  4.495925  6.471432  5.369522  7.459658 13.037437  2.995361
8    3.855261  3.855261  7.696892  3.211697  7.095775 11.195004  3.211697
9    1.974253  2.935926  7.100851 10.025427  5.820066  7.616000  2.935926
10   2.788086  1.974253  5.103091  1.974253  5.783481 10.940852  1.974253
    PGLYRP2  IGLV3.25  IGHV1.18 RHOXF1P3 RN7SL314P    LRP1B    NPY5R  IGLV3.10
1  1.974253  9.979446  8.058747 3.570125  1.974253 9.291321 7.510974  9.211595
2  3.375647  6.789203  7.358341 4.023423  1.974253 6.806549 9.303323  5.963283
3  2.561722 11.832746  9.333807 2.799514  1.974253 4.189189 8.677550 11.188431
4  2.616851  9.447326 10.460471 4.363248  1.974253 7.384237 1.974253  8.727555
5  5.444800  7.166306  7.080156 2.667333  1.974253 2.945428 2.667333  4.632252
6  4.123427  1.974253  3.287253 5.067778  1.974253 8.189192 3.776393  1.974253
7  5.295355 12.626540 13.258958 4.012144  2.703658 5.719941 7.434196 10.774767
8  4.768857  8.242015  7.642452 4.768857  2.862244 3.211697 8.680548  6.291424
9  3.701796 10.195541  8.264962 2.935926  1.974253 4.906282 4.158917 10.497123
10 7.349707 12.025430 11.142759 3.544867  1.974253 3.711402 7.088850 10.736734
        ROCR  IGLV6.57    GPR26    PEG10   MAB21L4       PI3     FABP4
1  11.730292  8.480065 3.886815 13.00660  6.522021 11.706382 12.239942
2   9.247330  7.029705 1.974253 10.92868 10.212251  7.800276  6.699192
3   8.211785 11.649990 1.974253 14.88960  6.096049  3.126459  7.488021
4  10.566059 10.443329 1.974253 12.94802  4.802748  3.490572  3.230074
5   8.263584  6.334417 4.177786 15.54072  2.667333  7.722162  7.960392
6   3.557730  9.486094 3.287253 11.54578  7.166762  8.455878 15.116330
7   8.045973 10.909376 3.212737  9.82102  5.905077  9.598525 12.247841
8   9.824186  7.608772 4.388685 16.16185  4.494116  4.683241  9.502314
9   8.738712 10.630045 3.904923 16.13972  4.906282  3.807544  8.977788
10  6.930682 10.551684 3.857646 12.80889  4.928896  6.569830  9.434904
        MSLN   IGLV3.1   IGHV1.2   ABCC12    PHGR1     CST4      DSG1     CDH19
1   6.667732 11.147370  8.325179 4.474662 1.974253 2.653211  6.177018 11.282368
2   5.149006  6.890274  5.092901 3.573645 1.974253 3.375647  6.366212 10.478568
3   9.314479 13.710699  6.397500 1.974253 2.561722 4.929414  4.830579  6.474694
4  10.345662  9.598040  7.958385 9.972732 2.875813 3.230074  8.199201 10.677068
5   5.096563  6.549419  5.724173 6.213501 1.974253 3.470948 13.742397  8.622866
6  10.683241  4.724912  3.557730 1.974253 2.918135 6.240718  3.961746  6.206986
7  14.978597 14.446139 11.731177 1.974253 3.391029 4.353367 10.049099  8.268723
8   6.995038  9.062704  4.591945 3.855261 2.862244 4.010580 10.759051  1.974253
9   9.634691 12.415763  8.436976 1.974253 1.974253 2.660432 14.275276  9.222733
10 13.354683 12.639260 10.933294 2.788086 2.788086 3.988488  6.803650  5.047398
       FOLR1      FGG   HORMAD1      LEP    NBPF4     IL20    CXCL17  IGHV3.48
1   5.424804 2.925978  9.772767 5.721759 5.967415 3.886815  7.773201  6.856367
2  10.638270 3.375647 11.854691 3.573645 3.573645 4.620609 10.145745  4.143545
3   3.733763 1.974253 11.241746 4.391992 4.189189 2.799514  6.216633  8.326088
4  10.537880 1.974253 10.681252 3.368196 2.875813 3.962139 14.512592 10.697179
5  10.379948 4.098072 10.787087 3.153202 3.324013 3.153202 12.211768  6.195365
6   4.818868 3.776393  7.677456 9.642129 1.974253 3.557730 12.454877  3.557730
7  12.700229 3.800333  9.096671 9.382679 3.212737 2.995361  6.154840 12.177050
8  11.452338 1.974253 10.079009 4.494116 3.211697 1.974253  4.010580  6.502117
9   7.210868 2.660432 10.263027 6.035425 1.974253 3.141857  3.585824  7.687497
10 12.029423 2.788086 10.093687 4.316298 4.316298 5.103091 12.940301  9.991861
    IGHV3.21   SCGB3A1    ABCC8      LCN2     EDN3    PTPRZ1    EEF1A2
1   7.468260  3.297826 6.878298  9.620247 4.968043 12.885746  4.350365
2   6.053438  3.742617 2.804049  8.805439 8.592228  9.032236  8.774958
3  12.454858  3.471244 1.974253  4.830579 3.809820  7.700018  3.565480
4   9.387125  3.230074 1.974253 12.081285 2.616851 13.224475  6.984291
5   5.505236  8.862352 4.738373  7.860613 6.549419  8.491916  3.922175
6   2.918135  8.637325 5.724588 10.661673 6.619470 10.870812 14.261109
7  13.074458  7.734556 6.402974 10.867209 5.802239  8.244521  6.112549
8   7.174654  7.696892 5.066653  9.838781 6.173221  9.970532  7.143619
9   9.011096  3.807544 1.974253  6.550590 9.378096  9.336254  4.992253
10 10.648329 14.287719 4.107143  9.899854 5.943209  5.783481  4.579104
   IGHV1.69D      DSC3   CYP2A7    HOTAIR    MUC19   PLA2G2A      KRT4     SIM2
1   7.830525 11.471231 2.653211  7.984931 2.925978 10.451661  7.198923 3.442332
2   5.445571  7.851266 2.804049  7.555754 1.974253  4.537741  1.974253 9.315476
3  10.274380  8.774323 2.799514  7.605276 1.974253  4.438228  3.809820 2.561722
4   9.326085 14.163417 1.974253  5.168309 7.605945  6.101341  3.368196 4.922790
5   5.246304 13.421454 4.252934 10.537198 3.470948  8.605850  3.153202 9.809162
6   1.974253 11.037975 3.287253  7.268176 3.287253 14.323387 11.651939 9.102821
7  13.051976 13.339559 3.391029  8.478326 2.703658 10.518415  4.495925 7.971248
8   6.203697 10.712262 1.974253  9.238108 5.471813  8.292375  1.974253 9.381245
9   8.140806 14.001415 2.660432  7.984448 2.935926  5.953129  1.974253 8.864600
10  7.852891  5.441324 3.110882  9.190526 4.656516  8.836603  2.788086 4.107143
      CLIC6      IGHM AC078993.1    TNNT1 LINC02562     CIDEA  IGHV5.51
1  7.099773 11.284420   1.974253 4.791236  2.653211  5.987789  8.845197
2  5.763899 10.047164   1.974253 6.661550  6.192133  3.375647  5.445571
3  6.891242 14.517448   2.978430 4.344109  1.974253  4.074757 13.153646
4  9.331419 11.615398   3.794802 5.534769  4.109951  1.974253 10.100075
5  6.414947 10.309182   2.667333 8.499307  6.619025  3.922175  5.868806
6  7.904353  7.713851   1.974253 9.805575  8.397419 10.424328  2.918135
7  9.293498 16.104628   4.193801 4.012144  2.995361  8.624265 12.221118
8  6.262774 11.809871   4.274274 7.620086  3.468942  2.862244  7.848870
9  6.419073 15.753530   2.660432 7.468796  1.974253  4.992253 10.120052
10 9.487459 14.477489   3.110882 2.788086  7.907927  5.258053  9.932444
      BRINP2       ER
1   7.707396 Negative
2  12.451769 Negative
3  11.452630 Negative
4   2.616851 Negative
5  11.089642 Negative
6   3.287253 Negative
7   5.473967 Negative
8   6.400630 Negative
9   3.807544 Negative
10  2.788086 Negative</code></pre>
</div>
</div>
<p>With the same reasoning as creating our initial training and testing set, none of the samples included in the training set at each split is found in the corresponding testing set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">intersect</span>(<span class="fu">training</span>(er_folds<span class="sc">$</span>splits[[<span class="dv">1</span>]])<span class="sc">$</span>ID, <span class="fu">testing</span>(er_folds<span class="sc">$</span>splits[[<span class="dv">1</span>]])<span class="sc">$</span>ID)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>character(0)</code></pre>
</div>
</div>
</section>
<section id="model-specifiction-and-recipe" class="level2">
<h2 class="anchored" data-anchor-id="model-specifiction-and-recipe">Model specifiction and recipe</h2>
<p>Now when we create our model specification we declare that <code>penalty</code> and <code>mixture</code> are going to be tuned. Our “recipe” uses the same formula as before, but need to be careful as we introduce an ID column into the data. The workflow is also created at this stage.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>lasso_spec_tune <span class="ot">&lt;-</span> <span class="fu">logistic_reg</span>(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">penalty =</span> <span class="fu">tune</span>(), <span class="co">#Penalty (lambda): Use tune() to test different shrinkage strengths.</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>     <span class="co"># Mixture (alpha): Use tune() to test the balance between Lasso (1) and Ridge (0) regularization, defining the Elastic Net.</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">mixture =</span> <span class="fu">tune</span>() </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_engine</span>(<span class="st">"glmnet"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_mode</span>(<span class="st">"classification"</span>)</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>er_recipe_multigene <span class="ot">&lt;-</span> <span class="fu">recipe</span>(ER <span class="sc">~</span> ., <span class="at">data =</span> er_train)  <span class="sc">%&gt;%</span> </span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">update_role</span>(ID, <span class="at">new_role =</span> <span class="st">"ID"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_normalize</span>(<span class="fu">all_numeric_predictors</span>())  </span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>   <span class="do">## Need to say that our ID column is an identifier - not something to be included in the model</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>elastic_net_wflow <span class="ot">&lt;-</span> <span class="fu">workflow</span>() <span class="sc">%&gt;%</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_recipe</span>(er_recipe_multigene) <span class="sc">%&gt;%</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_model</span>(lasso_spec_tune)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="setup-a-grid-for-tuning" class="level2">
<h2 class="anchored" data-anchor-id="setup-a-grid-for-tuning">Setup a grid for tuning</h2>
<p>Unfortunately there are an <em>infinite</em> number of possibilities that both these arguments can take and we clearly cannot test them all. What we do instead is try and explore the full range of values, and find roughly the region where the optimal values are located. In practical terms this is often perfectly adequate.</p>
<p>Infrastructure for tuning parameters is support by <code>dials</code> and <code>tune</code> which are both part of the <code>tidymodels</code> eco-system. i.e.&nbsp;they get installed and loaded when you install and load <code>tidymodels</code>.</p>
<ul>
<li><a href="https://dials.tidymodels.org/" class="uri">https://dials.tidymodels.org/</a></li>
<li><a href="https://tune.tidymodels.org/" class="uri">https://tune.tidymodels.org/</a></li>
</ul>
<p>There are a couple of convenient helper functions <code>penalty</code> and <code>mixture</code> that will create templates for choosing values for tuning. We then have to create a <em>grid</em> of values that we want to test. This is simply a data frame (/tibble) containing combinations of <code>penalty</code> and <code>mixture</code> and can be created with the <code>grid_regular</code> function. This function will ensure that the full range of values is tested. In other words that the maximum and minimum values are included in the search. With the <code>levels</code> argument we can control how finely we split the ranges.</p>
<p>Another option is to use <code>grid_random</code> to create a <em>random</em> set of combinations of the parameters. Instead of <code>levels</code> we say how many combinations we want in total using <code>size</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>penalty_range <span class="ot">&lt;-</span> <span class="fu">penalty</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">5</span>, <span class="dv">0</span>)) <span class="co"># use helper function to define range of penalty values</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>mixture_range <span class="ot">&lt;-</span> <span class="fu">mixture</span>() <span class="co"># use helper function to define a range of mixture values</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>elastic_net_grid_regular <span class="ot">&lt;-</span> <span class="fu">grid_regular</span>(</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  penalty_range, </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  mixture_range,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="at">penalty =</span> <span class="dv">5</span>, <span class="at">mixture =</span> <span class="dv">3</span>) <span class="co"># Test 5 penalty values x 3 mixture values = 15 combinations</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>elastic_net_grid_random <span class="ot">&lt;-</span> <span class="fu">grid_random</span>(</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  penalty_range, </span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  mixture_range,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">size =</span> <span class="dv">15</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With the setup we have just created there are 15 different models to be fit (one model for each combination in our grid), and the 10-fold strategy we defined with <code>vfolds_cv</code> means we have to fit each of these 15 models 10 times. In total we have 150 models, and for each we want to assess the performance. This is quite a lot of work! Fortunately we can do this quite efficiently with <code>tidymodels</code> and get an output that we can assess. The <code>tune_grid</code> function will use the cross-validation strategy defined with <code>vfolds_cv</code>, a grid structure and compute performance metrics for your workflow.</p>
</section>
<section id="inspect-tuning-results" class="level2">
<h2 class="anchored" data-anchor-id="inspect-tuning-results">Inspect tuning results</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>elastic_net_tune_res <span class="ot">&lt;-</span> <span class="fu">tune_grid</span>(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  elastic_net_wflow, </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">resamples =</span> er_folds, </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">grid =</span> elastic_net_grid_regular,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">metrics =</span> <span class="fu">metric_set</span>(accuracy, specificity, sensitivity, roc_auc)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The actual object itself is not particularly informative when printed:-</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>elastic_net_tune_res </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># Tuning results
# 10-fold cross-validation using stratification 
# A tibble: 10 × 4
   splits           id     .metrics          .notes          
   &lt;list&gt;           &lt;chr&gt;  &lt;list&gt;            &lt;list&gt;          
 1 &lt;split [761/86]&gt; Fold01 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;
 2 &lt;split [761/86]&gt; Fold02 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;
 3 &lt;split [761/86]&gt; Fold03 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;
 4 &lt;split [762/85]&gt; Fold04 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;
 5 &lt;split [763/84]&gt; Fold05 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;
 6 &lt;split [763/84]&gt; Fold06 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;
 7 &lt;split [763/84]&gt; Fold07 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;
 8 &lt;split [763/84]&gt; Fold08 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;
 9 &lt;split [763/84]&gt; Fold09 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;
10 &lt;split [763/84]&gt; Fold10 &lt;tibble [60 × 6]&gt; &lt;tibble [0 × 3]&gt;</code></pre>
</div>
</div>
<p>There is however a handy plotting method that will give a quick overview of the metrics achieved for different runs of the model. On the whole, it looks like a mixture of 0 is best with little difference as the penalty (amount of regularization) changes. Recall that a mixture of 0 corresponds to <em>Ridge</em>, as opposed to <em>Lasso</em> meaning that the final model that we fit will not set any coefficients to zero.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>elastic_net_tune_res <span class="sc">%&gt;%</span> autoplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/plot_tune_res-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also request details of the “best” models according to a particular metric. ROC AUC is generally a good idea as it balances the specificity and sensitivity. The metrics shown in the table are averaged over the v-folds, which is 10 in our case.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_best</span>(elastic_net_tune_res, <span class="at">metric =</span> <span class="st">"roc_auc"</span>, <span class="at">n =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 8
   penalty mixture .metric .estimator  mean     n std_err .config              
     &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;      &lt;dbl&gt; &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;                
1 0.0562         0 roc_auc binary     0.959    10 0.0103  Preprocessor1_Model04
2 1              0 roc_auc binary     0.958    10 0.00964 Preprocessor1_Model05
3 0.00001        0 roc_auc binary     0.956    10 0.0107  Preprocessor1_Model01
4 0.000178       0 roc_auc binary     0.956    10 0.0107  Preprocessor1_Model02
5 0.00316        0 roc_auc binary     0.956    10 0.0107  Preprocessor1_Model03</code></pre>
</div>
</div>
<p>If we want to concentrate on a specific measure (such as sensitivity), we can look at that too.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_best</span>(elastic_net_tune_res, <span class="at">metric =</span> <span class="st">"sensitivity"</span>, <span class="at">n=</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 8
   penalty mixture .metric     .estimator  mean     n std_err .config           
     &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;      &lt;dbl&gt; &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;             
1 0.0562         0 sensitivity binary     0.855    10  0.0312 Preprocessor1_Mod…
2 0.00001        0 sensitivity binary     0.849    10  0.0348 Preprocessor1_Mod…
3 0.000178       0 sensitivity binary     0.849    10  0.0348 Preprocessor1_Mod…
4 0.00316        0 sensitivity binary     0.849    10  0.0348 Preprocessor1_Mod…
5 0.00316        1 sensitivity binary     0.818    10  0.0346 Preprocessor1_Mod…</code></pre>
</div>
</div>
</section>
<section id="create-final-model-and-fit" class="level2">
<h2 class="anchored" data-anchor-id="create-final-model-and-fit">Create final model and fit</h2>
<p>Once we are happy, we can save the details of the best model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>best_params <span class="ot">&lt;-</span> <span class="fu">select_best</span>(elastic_net_tune_res, <span class="at">metric =</span> <span class="st">"roc_auc"</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>best_params</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  penalty mixture .config              
    &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;                
1  0.0562       0 Preprocessor1_Model04</code></pre>
</div>
</div>
<p>And the choice of parameters can be fed back to our workflow. Our workflow up to this point only has placeholders for the <code>mixture</code> and <code>penalty</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>elastic_net_wflow</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>══ Workflow ════════════════════════════════════════════════════════════════════
Preprocessor: Recipe
Model: logistic_reg()

── Preprocessor ────────────────────────────────────────────────────────────────
1 Recipe Step

• step_normalize()

── Model ───────────────────────────────────────────────────────────────────────
Logistic Regression Model Specification (classification)

Main Arguments:
  penalty = tune()
  mixture = tune()

Computational engine: glmnet </code></pre>
</div>
</div>
<p>With <code>finalize_workflow</code> these values get updated according to our final choices.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>final_wflow <span class="ot">&lt;-</span> elastic_net_wflow <span class="sc">%&gt;%</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">finalize_workflow</span>(best_params)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>final_wflow</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>══ Workflow ════════════════════════════════════════════════════════════════════
Preprocessor: Recipe
Model: logistic_reg()

── Preprocessor ────────────────────────────────────────────────────────────────
1 Recipe Step

• step_normalize()

── Model ───────────────────────────────────────────────────────────────────────
Logistic Regression Model Specification (classification)

Main Arguments:
  penalty = 0.0562341325190349
  mixture = 0

Computational engine: glmnet </code></pre>
</div>
</div>
<p>Remember that the purpose of tuning was to find an optimal set of hyperparameters, and we did that using a newly-created set of cross-validation subsets. We still need to fit, our create, the final model using those paramaters that we chose. This still needs to be performed on our entire training dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>final_model_fit <span class="ot">&lt;-</span> <span class="fu">fit</span>(final_wflow, <span class="at">data =</span> er_train)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>final_model_fit <span class="sc">%&gt;%</span> tidy</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 246 × 3
   term        estimate penalty
   &lt;chr&gt;          &lt;dbl&gt;   &lt;dbl&gt;
 1 (Intercept)  2.71     0.0562
 2 CLEC3A       0.0756   0.0562
 3 SCGB2A2     -0.0580   0.0562
 4 CPB1         0.0248   0.0562
 5 GSTM1       -0.204    0.0562
 6 TFF1         0.116    0.0562
 7 S100A7      -0.0151   0.0562
 8 SCGB1D2      0.00630  0.0562
 9 KCNJ3        0.0352   0.0562
10 LINC00993   -0.0425   0.0562
# ℹ 236 more rows</code></pre>
</div>
</div>
<p>Finally we can make predictions using our testing data <strong>which hasn’t been used in any of the previous steps</strong> and assess the metrics</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>class_metrics <span class="ot">&lt;-</span> <span class="fu">metric_set</span>(accuracy, sensitivity, specificity)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(er_test ,ER) <span class="sc">%&gt;%</span> </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(<span class="fu">predict</span>(final_model_fit, <span class="at">new_data =</span> er_test)) <span class="sc">%&gt;%</span> </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class_metrics</span>(<span class="at">truth =</span> ER, <span class="at">estimate =</span> .pred_class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
  .metric     .estimator .estimate
  &lt;chr&gt;       &lt;chr&gt;          &lt;dbl&gt;
1 accuracy    binary         0.925
2 sensitivity binary         0.918
3 specificity binary         0.927</code></pre>
</div>
</div>
<p>The accuracy, sensitivity and specificity from our previous logistic regression were 0.920, 0.857 and 0.939 respectively so it looks like we are improving our sensitivity at the cost of a slight decrease in specificity. After tuning we have ended up with a model that is doing better at predicting both positive and negative cases .🎉</p>
</section>
<section id="training-for-random-forests" class="level2">
<h2 class="anchored" data-anchor-id="training-for-random-forests">Training for Random Forests</h2>
<p>We’ll now walk through an example of tuning a random forest, and hopefully you will see that the process is roughly the same. A useful starting point is the <code>show_model_info</code> function that will give us plenty of information about using a particular machine learning methods, the “engines” that can be used and what arguments can be set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_model_info</span>(<span class="st">"rand_forest"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Information for `rand_forest`
 modes: unknown, classification, regression, censored regression 

 engines: 
   classification: randomForest, ranger¹, spark
   regression:     randomForest, ranger¹, spark

¹The model can use case weights.

 arguments: 
   ranger:       
      mtry  --&gt; mtry
      trees --&gt; num.trees
      min_n --&gt; min.node.size
   randomForest: 
      mtry  --&gt; mtry
      trees --&gt; ntree
      min_n --&gt; nodesize
   spark:        
      mtry  --&gt; feature_subset_strategy
      trees --&gt; num_trees
      min_n --&gt; min_instances_per_node

 fit modules:
         engine           mode
         ranger classification
         ranger     regression
   randomForest classification
   randomForest     regression
          spark classification
          spark     regression

 prediction modules:
             mode       engine                    methods
   classification randomForest           class, prob, raw
   classification       ranger class, conf_int, prob, raw
   classification        spark                class, prob
       regression randomForest               numeric, raw
       regression       ranger     conf_int, numeric, raw
       regression        spark                    numeric</code></pre>
</div>
</div>
<p>In the output it lists <code>mtry</code>, <code>trees</code> and <code>min_n</code> all as arguments that can be tuned. These arguments control the diversity (<code>mtry</code>), complexity (<code>min_n</code>) and number of trees used for consensus voting (<code>trees</code>). The purpose of tuning these parameters is to give a selection of tree with enough complexity and diversity in our training data without overfitting. The particular optimal combination of these values with vary from dataset to dataset, and hence why they need to be tuned. A random forest specification with hyper-parameters to be tuned can be created as follows:-</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>rf_spec_tune <span class="ot">&lt;-</span> <span class="fu">rand_forest</span>(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">trees=</span><span class="fu">tune</span>(), </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">mtry =</span> <span class="fu">tune</span>(),</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_n =</span> <span class="fu">tune</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_engine</span>(<span class="st">"ranger"</span>, <span class="at">importance =</span> <span class="st">"permutation"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_mode</span>(<span class="st">"classification"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and now create the workflow</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>rf_workflow <span class="ot">&lt;-</span> <span class="fu">workflow</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_model</span>(rf_spec_tune) <span class="sc">%&gt;%</span> </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_recipe</span>(er_recipe_multigene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To define the grid for searching we can again make use of the <code>dials</code> package that had pre-defined notions of what values the <code>mtry</code>, <code>trees</code> and <code>min_n</code> should contain and stop us accidentally trying to create a model with invalid parameters values. This would be useful in situation where we were trying more advance methods of grid searching, or even using random grids.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>mtry_range <span class="ot">&lt;-</span> <span class="fu">mtry</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">50</span>))</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>tree_range <span class="ot">&lt;-</span> <span class="fu">trees</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">500</span>,<span class="dv">2000</span>))</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>minn_range <span class="ot">&lt;-</span> <span class="fu">min_n</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">50</span>)) </span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>rf_grid <span class="ot">&lt;-</span> <span class="fu">grid_regular</span>(</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>  mtry_range,</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  tree_range,</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>  minn_range,</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="at">mtry =</span> <span class="dv">5</span>, <span class="at">trees =</span> <span class="dv">3</span>, <span class="at">min_n =</span> <span class="dv">3</span>)</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>rf_grid</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 45 × 3
    mtry trees min_n
   &lt;int&gt; &lt;int&gt; &lt;int&gt;
 1     5   500     5
 2    16   500     5
 3    27   500     5
 4    38   500     5
 5    50   500     5
 6     5  1250     5
 7    16  1250     5
 8    27  1250     5
 9    38  1250     5
10    50  1250     5
# ℹ 35 more rows</code></pre>
</div>
</div>
<p>We are now ready to do the tuning, which without any further modifications to the code would take a <em>lot</em> longer. Fortunately, I have Dr.&nbsp;Lewis Quayle to thank for some code on his <a href="https://www.lewisdoesdata.com/2025/10/22/introduction-to-tidy-modelling-part-2.html">post on tidymodels</a> to configure <code>tidymodels</code> to use multiple CPU cores on your computer where available.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>We can run the code in parallel because the tuning process is inherently an example of what was called “embarrassingly parallel” back in the day (people sometimes prefer “Pleasingly parallel” nowadays). In other words, the training / testing occurring at each fold in our workflow is completely independant and can therefore be run at the same time if possible rather than waiting for the previous fold to finish.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="do">## setting a seed required for reproducibility</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="co"># register a parallel back-end</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>future<span class="sc">::</span><span class="fu">plan</span>(future<span class="sc">::</span><span class="fu">multisession</span>(), <span class="at">workers =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>() <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>rf_tune_res <span class="ot">&lt;-</span> <span class="fu">tune_grid</span>(</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>  rf_workflow, </span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">resamples =</span> er_folds, </span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">grid =</span> rf_grid,</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">metrics =</span> <span class="fu">metric_set</span>(accuracy, specificity, sensitivity, roc_auc)</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As before, we can plot to get a quick overview.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>rf_tune_res <span class="sc">%&gt;%</span> <span class="fu">autoplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/rf_tuning_auto-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<p>The plot chooses informative labels where “Minimal Node Size” = <code>min_n</code> and “# Randomly Selected Predictors” = <code>mtry</code>. If we wish to focus on Area Under the Curve (<code>roc_auc</code>) the best combination looks to be lower <code>min_n</code> and <code>mtry</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_best</span>(elastic_net_tune_res, <span class="at">metric =</span> <span class="st">"roc_auc"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 8
   penalty mixture .metric .estimator  mean     n std_err .config              
     &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;      &lt;dbl&gt; &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;                
1 0.0562         0 roc_auc binary     0.959    10 0.0103  Preprocessor1_Model04
2 1              0 roc_auc binary     0.958    10 0.00964 Preprocessor1_Model05
3 0.00001        0 roc_auc binary     0.956    10 0.0107  Preprocessor1_Model01
4 0.000178       0 roc_auc binary     0.956    10 0.0107  Preprocessor1_Model02
5 0.00316        0 roc_auc binary     0.956    10 0.0107  Preprocessor1_Model03</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>best_params <span class="ot">&lt;-</span> <span class="fu">select_best</span>(rf_tune_res, <span class="at">metric =</span> <span class="st">"roc_auc"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And now we can create the final model and check the importance of the features</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>rf_final_workflow <span class="ot">&lt;-</span> rf_workflow <span class="sc">%&gt;%</span> </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">finalize_workflow</span>(best_params)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>rf_final_fit <span class="ot">&lt;-</span> rf_final_workflow <span class="sc">%&gt;%</span> </span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fit</span>(er_train)</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>importance_df <span class="ot">&lt;-</span> <span class="fu">pull_workflow_fit</span>(rf_final_fit)<span class="sc">$</span>fit<span class="sc">$</span>variable.importance <span class="sc">%&gt;%</span></span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>(<span class="at">name =</span> <span class="st">"term"</span>, <span class="at">value =</span> <span class="st">"Importance"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(Importance))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and we can make a simple bar plot if we wish</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>importance_df <span class="sc">%&gt;%</span> </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="do">## pick the top 20 </span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(Importance, <span class="at">n =</span> <span class="dv">20</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> Importance, <span class="at">y =</span> forcats<span class="sc">::</span><span class="fu">fct_reorder</span>(term, Importance))) <span class="sc">+</span> <span class="fu">geom_col</span>(<span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Gene"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"Importance"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/plot_rf_importance-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>and the final predictions and evaluation follow a pattern that is hopefully familiar by now</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>rf_final_fit <span class="sc">%&gt;%</span> </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">predict</span>(<span class="at">new_data =</span>  er_test) <span class="sc">%&gt;%</span> </span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(<span class="fu">select</span>(er_test, ER)) <span class="sc">%&gt;%</span> </span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class_metrics</span>(<span class="at">truth =</span> ER, <span class="at">estimate =</span> .pred_class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
  .metric     .estimator .estimate
  &lt;chr&gt;       &lt;chr&gt;          &lt;dbl&gt;
1 accuracy    binary         0.925
2 sensitivity binary         0.898
3 specificity binary         0.933</code></pre>
</div>
</div>
</section>
<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>This concludes the end of this little series on using <code>tidymodels</code> for classification of omics data. I hope that I have covered the main principles of the <code>tidymodels</code> eco-system. Just looking through the code I can see a lot of places where the code is reused, which is good for readability and extendability. Even though I have not covered all the possible classification methods, it <em>should</em> be possible with relative ease (that is relative to coding everything from scratch) to translate what we have covered here to other methods. This is real advantage of <code>tidymodels</code> in that most of our time should be spent interpreting and refining that having to code everything by-hand. I haven’t even covered the use of regression vs classification (that is, predicting a <em>numeric</em> rather than categorical outcome) and again this should be within your reach now by using <code>set_mode("regression")</code> instead of <code>classification</code> in the models covered here.</p>
<p>Of course it would be remiss of me not to mention AI and deep learning methods that are increasingly being used in research. Whilst these are more complex methods, the core principles of training / testing and tuning are still highly applicable. Furthermore, these methods are still completely dependant on having appropriately filtered and cleaned data as a starting point. Although Python is probably the market leader for AI, R actually has interfaces to keras and tensorflow (two of the most popular implementations). There is no need to abandon R just yet.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>