<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mark Dunning">
<meta name="dcterms.date" content="2025-10-23">

<title>Introduction to RNA-Seq - Part 2 – Dr.&nbsp;Mark Dunning - Bioinformatician / Data Scientist</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-2fef5ea3f8957b3e4ecc936fc74692ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-1688d9e1630bd0bcb1ec498477fd8fb7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-96TB60ETKS"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-96TB60ETKS', { 'anonymize_ip': true});
</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Dr.&nbsp;Mark Dunning - Bioinformatician / Data Scientist</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../cv.html"> 
<span class="menu-text">CV</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../training.html"> 
<span class="menu-text">Training Materials</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../posts.html"> 
<span class="menu-text">Posts</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a>
  <ul class="collapse">
  <li><a href="#quick-start" id="toc-quick-start" class="nav-link" data-scroll-target="#quick-start">Quick Start</a></li>
  </ul></li>
  <li><a href="#differential-expression-with-deseq2" id="toc-differential-expression-with-deseq2" class="nav-link" data-scroll-target="#differential-expression-with-deseq2">Differential expression with <code>DESeq2</code></a>
  <ul class="collapse">
  <li><a href="#recap-of-pre-processing" id="toc-recap-of-pre-processing" class="nav-link" data-scroll-target="#recap-of-pre-processing">Recap of pre-processing</a></li>
  </ul></li>
  <li><a href="#the-deseq-workflow-in-brief" id="toc-the-deseq-workflow-in-brief" class="nav-link" data-scroll-target="#the-deseq-workflow-in-brief">The DESeq workflow in brief</a></li>
  <li><a href="#processing-the-de-results-using-tidyverse" id="toc-processing-the-de-results-using-tidyverse" class="nav-link" data-scroll-target="#processing-the-de-results-using-tidyverse">Processing the DE results using tidyverse</a></li>
  <li><a href="#changing-the-direction-of-the-contrast-or-the-groups-being-compared" id="toc-changing-the-direction-of-the-contrast-or-the-groups-being-compared" class="nav-link" data-scroll-target="#changing-the-direction-of-the-contrast-or-the-groups-being-compared">Changing the direction of the contrast, or the groups being compared</a>
  <ul class="collapse">
  <li><a href="#exercise" id="toc-exercise" class="nav-link" data-scroll-target="#exercise">Exercise</a></li>
  </ul></li>
  <li><a href="#more-complex-designs" id="toc-more-complex-designs" class="nav-link" data-scroll-target="#more-complex-designs">More complex designs</a></li>
  <li><a href="#adding-annotation-to-the-deseq2-results" id="toc-adding-annotation-to-the-deseq2-results" class="nav-link" data-scroll-target="#adding-annotation-to-the-deseq2-results">Adding annotation to the DESeq2 results</a>
  <ul class="collapse">
  <li><a href="#exercise-1" id="toc-exercise-1" class="nav-link" data-scroll-target="#exercise-1">Exercise</a></li>
  <li><a href="#the-volcano-plot" id="toc-the-volcano-plot" class="nav-link" data-scroll-target="#the-volcano-plot">The volcano plot</a>
  <ul class="collapse">
  <li><a href="#exporting-normalized-counts" id="toc-exporting-normalized-counts" class="nav-link" data-scroll-target="#exporting-normalized-counts">Exporting normalized counts</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#full-deseq-workflow" id="toc-full-deseq-workflow" class="nav-link" data-scroll-target="#full-deseq-workflow">Full DESeq workflow</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Introduction to RNA-Seq - Part 2</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mark Dunning </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 23, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="overview" class="level1">
<h1>Overview</h1>
<p>Performing differential expression on Bulk RNA-seq data using DESeq2</p>
<section id="quick-start" class="level2">
<h2 class="anchored" data-anchor-id="quick-start">Quick Start</h2>
<p>This section follows on from <a href="../..\training/bulk-rnaseq_part1/index.html">Part 1</a> where we saw how to import raw RNA-seq counts into <code>DESeq2</code> and perform some quality assessment. Several packages are required, which can be downloaded with this code:-</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"https://raw.githubusercontent.com/markdunning/markdunning.github.com/refs/heads/master/files/training/bulk_rnaseq/install_bioc_packages.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The following will also assume you have created a <code>DESeq2</code> object in a folder called <code>Robjects</code> in your working directory. This can be downloaded with the following.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"Robjects/"</span>,<span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://github.com/markdunning/markdunning.github.com/raw/refs/heads/master/files/training/bulk_rnaseq/dds.rds"</span>,<span class="at">destfile =</span> <span class="st">"Robjects/dds.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="differential-expression-with-deseq2" class="level1">
<h1>Differential expression with <code>DESeq2</code></h1>
<p>Now that we are happy that the data quality looks good, we can proceed to test for differentially expressed genes. There are a number of packages to analyse RNA-Seq data. Most people use <code>DESeq2</code>, <code>edgeR</code> or <code>limma</code>. We will use <code>DESeq2</code> for the rest of this practical.</p>
<section id="recap-of-pre-processing" class="level2">
<h2 class="anchored" data-anchor-id="recap-of-pre-processing">Recap of pre-processing</h2>
<p>The previous section walked-through the pre-processing and transformation of the count data. Here, for completeness, we list the minimal steps required to process the data prior to differential expression analysis.</p>
<p>Note that although we spent some time looking at the quality of our data , these steps are not strictly <em>necessary</em> prior to performing differential expression so are not shown here for the sake of brevity. Remember, <code>DESeq2</code> <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts">requires raw counts</a> so the <code>vst</code> transformation is not shown as part of this basic protocol.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>count_file <span class="ot">&lt;-</span> <span class="st">"raw_counts/raw_counts_matrix.tsv"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(count_file)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Step needed for this data to tidy the column names</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(counts)[<span class="sc">-</span><span class="dv">1</span>] <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_remove_all</span>(<span class="fu">colnames</span>(counts)[<span class="sc">-</span><span class="dv">1</span>], <span class="st">"X"</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>sampleinfo_corrected <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"meta_data/sampleInfo_corrected.txt"</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(counts, </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>                                <span class="at">colData =</span> sampleinfo_corrected,</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>                                <span class="at">design =</span> <span class="sc">~</span>condition, <span class="at">tidy=</span><span class="cn">TRUE</span>)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(dds, <span class="at">file=</span><span class="st">"Robjects/dds.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will be using these raw counts throughout the workshop and transforming them using methods in the <code>DESeq2</code> package. If you want to know about alternative methods for count normalisation they are covered on <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">this page</a>.</p>
</section>
</section>
<section id="the-deseq-workflow-in-brief" class="level1">
<h1>The DESeq workflow in brief</h1>
<p>We have previously defined the <code>condition</code> as our factor of interest using the <code>design</code> argument when we created the object. This can be checked using the <code>design</code> function. The</p>
<p>The current design can be retrieved using the <code>design</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">design</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>~condition</code></pre>
</div>
</div>
<p>It can be changed at any point before we run the differential expression workflow, can refer to any of the variables saved in the <code>colData</code> of the <code>dds</code> object; or even a combination of different variables.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 9 rows and 5 columns
                     Run condition        Name Replicate     Treated
             &lt;character&gt;  &lt;factor&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt;
1_CTR_BC_2   1_CTR_BC_2        CTR       CTR_1         1           N
2_TGF_BC_4    2_TGF_BC_4       TGF       TGF_1         1           Y
3_IR_BC_5      3_IR_BC_5       IR         IR_1         1           Y
4_CTR_BC_6    4_CTR_BC_6       CTR       CTR_2         2           N
5_TGF_BC_7    5_TGF_BC_7       TGF       TGF_2         2           Y
6_IR_BC_12    6_IR_BC_12       IR         IR_2         2           Y
7_CTR_BC_13 7_CTR_BC_13        CTR       CTR_3         3           N
8_TGF_BC_14  8_TGF_BC_14       TGF       TGF_3         3           Y
9_IR_BC_15    9_IR_BC_15       IR         IR_3         3           Y</code></pre>
</div>
</div>
<p>The counts that we have obtained via sequencing are subject to random sources of variation. The purpose of differential expression is to determine if potential sources of biological variation (e.g.&nbsp;counts observed from different sample groups) are greater than random noise.</p>
<p>The <code>DESeq</code> function runs a couple of processing steps automatically to adjust for different library size and gene-wise variability, which you can read about in the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#the-deseq2-model">DESeq2 vignette</a> and run some example code at the <a href="#full-deseq-workflow">end of this session</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>de_condition <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>de_condition</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 57914 9 
metadata(1): version
assays(4): counts mu H cooks
rownames(57914): ENSG00000000003 ENSG00000000005 ... ENSG00000284747
  ENSG00000284748
rowData names(26): baseMean baseVar ... deviance maxCooks
colnames(9): 1_CTR_BC_2 2_TGF_BC_4 ... 8_TGF_BC_14 9_IR_BC_15
colData names(6): Run condition ... Treated sizeFactor</code></pre>
</div>
</div>
<p>The results of the analysis are not immediately accessible, but can be obtained using the <code>results</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(de_condition)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): condition TGF vs CTR 
Wald test p-value: condition TGF vs CTR 
DataFrame with 57914 rows and 6 columns
                   baseMean log2FoldChange     lfcSE       stat     pvalue
                  &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;
ENSG00000000003 1430.562846     -0.2562067 0.0813936 -3.1477521 0.00164531
ENSG00000000005    0.113566     -0.0883933 4.0804559 -0.0216626 0.98271709
ENSG00000000419 1790.537536     -0.1449058 0.1008195 -1.4372794 0.15063862
ENSG00000000457  640.692302     -0.3106360 0.1137555 -2.7307347 0.00631933
ENSG00000000460  206.179026      0.1600274 0.1567963  1.0206074 0.30744046
...                     ...            ...       ...        ...        ...
ENSG00000284744    8.307038      -0.209531  0.800702  -0.261684   0.793565
ENSG00000284745    0.000000             NA        NA         NA         NA
ENSG00000284746    0.101097      -1.050150  4.080456  -0.257361   0.796900
ENSG00000284747   28.783710      -0.297530  0.448931  -0.662751   0.507490
ENSG00000284748    0.548323       0.873372  4.080456   0.214038   0.830517
                      padj
                 &lt;numeric&gt;
ENSG00000000003 0.00906995
ENSG00000000005         NA
ENSG00000000419 0.30807839
ENSG00000000457 0.02685529
ENSG00000000460 0.50350287
...                    ...
ENSG00000284744         NA
ENSG00000284745         NA
ENSG00000284746         NA
ENSG00000284747   0.689074
ENSG00000284748         NA</code></pre>
</div>
</div>
<p>Each row is a particular gene measured in the study (i.e.&nbsp;all genes in the organism being studied) and each column reports some aspect of the differential expression analysis for that gene. The values that I typically pay most attention to are highlighted in <strong>bold</strong>.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 30%">
<col style="width: 38%">
<col style="width: 30%">
</colgroup>
<thead>
<tr class="header">
<th>Column Name</th>
<th>Description</th>
<th>Importance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>baseMean</strong></td>
<td>The average of normalized count values (size-factor corrected) across all samples in the comparison.</td>
<td>Measures the average expression magnitude of the gene and useful for basic filtering</td>
</tr>
<tr class="even">
<td><strong>log2FoldChange</strong></td>
<td>The estimated log base 2 fold change between the two groups being contrasted (e.g., Treatment / Control).</td>
<td>The primary measure of the effect size or magnitude of the difference.</td>
</tr>
<tr class="odd">
<td>lfcSE</td>
<td>The standard error estimate for the <span class="math inline">\(\text{log}_2\text{FoldChange}\)</span>.</td>
<td>Measures the precision of the <span class="math inline">\(\text{log}_2\text{FC}\)</span> estimate. Smaller values indicate higher reliability.</td>
</tr>
<tr class="even">
<td>stat</td>
<td>The Wald test statistic (calculated as <span class="math inline">\(\text{log}_2\text{FC} / \text{lfcSE}\)</span>)</td>
<td>The statistical value used to test the null hypothesis (that the <span class="math inline">\(\text{log}_2\text{FC}\)</span> is zero)</td>
</tr>
<tr class="odd">
<td>pvalue</td>
<td>The <span class="math inline">\(p\)</span>-value derived from the Wald test statistic.</td>
<td>The unadjusted probability of observing the effect by chance. Do not use this for filtering.</td>
</tr>
<tr class="even">
<td><strong>padj</strong></td>
<td>The Benjamini-Hochberg adjusted <span class="math inline">\(p\)</span>-value (False Discovery Rate, FDR).CRITICAL for significance. This corrects for multiple testing</td>
<td>CRITICAL for statistical significance. This corrects for multiple testing. Genes are considered differentially expressed if this value is below your threshold (e.g., 0.05).</td>
</tr>
</tbody>
</table>
<p>Note that <strong>all genes</strong> are reported. At this stage the gene identifiers are not very informative, something we will <a href="#adding-annotation-to-the-deseq2-results">fix shortly</a>. Furthermore, the <code>results</code> function displays results in a format which is not compatible with standard data manipulation tools (i.e.&nbsp;<code>tidyverse</code>), so we will have to convert.</p>
</section>
<section id="processing-the-de-results-using-tidyverse" class="level1">
<h1>Processing the DE results using tidyverse</h1>
<p>The output can be converted into a data frame and manipulated in the usual manner if we add the <code>tidy = TRUE</code> argument to <code>results</code>. It is recommended to use <code>dplyr</code> to manipulate the data frames with the standard set of operations detailed on the <a href="https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf">dplyr cheatsheet</a></p>
<ul>
<li><code>select</code> to pick which columns to display</li>
<li><code>filter</code> to restrict the rows</li>
<li><code>mutate</code> to add new variables to the data frame</li>
<li><code>arrange</code> to order the data frame according to values of a column or columns</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>%&gt;%</code> symbol refers to the piping operationin R, which is a way of chaining operations together. This was described in the introductory R notes found else on my pages</p>
<ul>
<li><a href="https://mdbioinformatics.com/training/r_part2/#piping" class="uri">https://mdbioinformatics.com/training/r_part2/#piping</a></li>
</ul>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">results</span>(de_condition, <span class="at">tidy=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">slice_head</span>(<span class="at">n=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               row     baseMean log2FoldChange      lfcSE       stat
1  ENSG00000000003 1430.5628457    -0.25620674 0.08139356 -3.1477521
2  ENSG00000000005    0.1135661    -0.08839330 4.08045587 -0.0216626
3  ENSG00000000419 1790.5375358    -0.14490579 0.10081950 -1.4372794
4  ENSG00000000457  640.6923020    -0.31063596 0.11375545 -2.7307347
5  ENSG00000000460  206.1790264     0.16002744 0.15679627  1.0206074
6  ENSG00000000938    7.4126750    -1.94966011 0.94841767 -2.0556978
7  ENSG00000000971 5318.4200170    -0.37348479 0.06859390 -5.4448693
8  ENSG00000001036 3629.7262081    -0.42700955 0.05618654 -7.5998545
9  ENSG00000001084  923.1202025    -0.19549932 0.09925106 -1.9697455
10 ENSG00000001167 1261.1547094     0.08263564 0.09065646  0.9115251
         pvalue         padj
1  1.645312e-03 9.069945e-03
2  9.827171e-01           NA
3  1.506386e-01 3.080784e-01
4  6.319333e-03 2.685529e-02
5  3.074405e-01 5.035029e-01
6  3.981166e-02           NA
7  5.184345e-08 9.308987e-07
8  2.964639e-14 1.176542e-12
9  4.886755e-02 1.346109e-01
10 3.620188e-01 5.605916e-01</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Although, the “tidy” output is more useful for downstream analysis, I personally like the output generated by the <code>results</code> function without <code>tidy=TRUE</code> as it displays the contrast being tested (in this case <code>TGF vs CTR</code>). This is a good way of checking that you are comparing the groups that you expect.</p>
</div>
</div>
<p>We can sort the rows by adjusted p-value and then print the first 10 rows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(de_condition,<span class="at">tidy=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(padj) <span class="sc">%&gt;%</span>  </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               row    baseMean log2FoldChange      lfcSE     stat        pvalue
1  ENSG00000119681   6471.7058       2.372065 0.07076496 33.52033 2.437260e-246
2  ENSG00000120708  22001.6982       1.436637 0.05904590 24.33084 9.248277e-131
3  ENSG00000172061   1481.6147       2.384971 0.09843523 24.22883 1.105455e-129
4  ENSG00000049540    932.0366       3.999465 0.17556867 22.78006 7.228861e-115
5  ENSG00000060718   1512.0092       2.123432 0.09532775 22.27506 6.448803e-110
6  ENSG00000115414 683348.5240       1.453310 0.06867840 21.16109  2.181368e-99
7  ENSG00000187498  30609.2525       1.458950 0.06926943 21.06196  1.776880e-98
8  ENSG00000186340  14380.2024       1.468632 0.06999698 20.98136  9.707600e-98
9  ENSG00000087245  51862.2917       1.282562 0.06613894 19.39194  9.026914e-84
10 ENSG00000139211   2302.3470       2.070441 0.10902113 18.99119  2.017183e-80
            padj
1  4.371957e-242
2  8.294780e-127
3  6.609883e-126
4  3.241783e-111
5  2.313573e-106
6   6.521564e-96
7   4.553383e-95
8   2.176687e-94
9   1.799164e-80
10  3.618422e-77</code></pre>
</div>
</div>
<p>Or we can sort the rows and then write the resulting data frame to a file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"de_analysis"</span>,<span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">results</span>(de_condition,<span class="at">tidy=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(padj) <span class="sc">%&gt;%</span> </span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>   readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="st">"de_analysis/condition_TGF_vs_CTR_DESeq_all.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Filtering to the differentially-expressed genes can be achieved using the <code>filter</code> function from <code>dplyr</code> with some cut-off on the adjusted p-value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">results</span>(de_condition,<span class="at">tidy=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="st">"de_analysis/condition_TGF_vs_CTR_DESeq_sig.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is also a good idea to save the DESeq output object itself so we can re-use later.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(de_condition, <span class="at">file=</span><span class="st">"Robjects/de_condition.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can discover how many differentially-expressed genes (at a particular p-value cut-off) using the <code>count</code> function. I am making sure that I explictly use the <code>count</code> function from <code>dplyr</code> using the technique of putting <code>dplyr::</code> before <code>count</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(de_condition,<span class="at">tidy=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">count</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  padj &lt; 0.05     n
1       FALSE 13039
2        TRUE  4899
3          NA 39976</code></pre>
</div>
</div>
<p>Another overview of the results is to use the <code>plotMA</code> function. Each point on this plot represents and individual gene with the x- and y-axes being the overall expression level and magnitude of difference respectively. Statistically significant genes are automatically highlighted. The fanning effect at low expression levels is often seen due to high relative fold-change at low expression levels. Once you have run your DE analysis, an MA plot visually confirms that your DE genes are not exclusively confined to extremely low-count regions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMA</span>(de_condition)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It is also instructive to perform a “sanity” check and plot the sample-level counts for genes with high significance. This could highlight any other technical factors that we are not currently taking into account. The plot is not particularly attractive, but is a good quick diagnostic as it confirms which direction your contrast is. i.e.&nbsp;Treatment vs Control or Control vs Treatment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>top_up <span class="ot">&lt;-</span> <span class="fu">results</span>(de_condition, <span class="at">tidy =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(log2FoldChange <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_min</span>(padj) <span class="sc">%&gt;%</span> </span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(row)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plotCounts</span>(dds,top_up,<span class="at">intgroup =</span> <span class="st">"condition"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>top_down <span class="ot">&lt;-</span> <span class="fu">results</span>(de_condition, <span class="at">tidy =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(log2FoldChange <span class="sc">&lt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_min</span>(padj) <span class="sc">%&gt;%</span> </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(row)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plotCounts</span>(dds,top_down,<span class="at">intgroup =</span> <span class="st">"condition"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If your study involves knocking-out a particular gene, or you have some positive controls that are known in advance, it would be a good idea to visualise their expression level with <code>plotCounts</code>.</p>
</div>
</div>
</section>
<section id="changing-the-direction-of-the-contrast-or-the-groups-being-compared" class="level1">
<h1>Changing the direction of the contrast, or the groups being compared</h1>
<p>You hopefully noticed that the sign of the log fold changes have been calculated using the <code>CTR</code> group as a baseline. In other words, a positive <code>log2FoldChange</code> means that a gene has higher expression in <code>TGF</code> compared to <code>CTR</code>. This is usually the direction that makes sense for biological interpretation. However, <code>DESeq2</code> has not done anything clever here. The order of the contrasts is dictated by the “levels” of the variable used in the design.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>dds<span class="sc">$</span>condition</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] CTR TGF IR  CTR TGF IR  CTR TGF IR 
Levels: CTR IR TGF</code></pre>
</div>
</div>
<p>Since <code>CTR</code> is the first thing printed here it gets chosen as the baseline. It is just a fortunate coincidence that <code>CTR</code> is first alphabetically! In order that we are not relying on the defaults and for transparency, I tend to explicitly state which contrast I want to make using the <code>contrast</code> argument and have full control over what the baseline is. In the <code>contrast</code> argument you create a vector (using <code>c</code>) followed by the levels in the order you want to compare with the baseline coming last.</p>
<p>In the below this can be read as <code>TGF</code> vs <code>CTR</code>, which is in fact the default</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This should give the same as the table above</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(de_condition, <span class="at">contrast=</span><span class="fu">c</span>(<span class="st">"condition"</span>,<span class="st">"TGF"</span>,<span class="st">"CTR"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): condition TGF vs CTR 
Wald test p-value: condition TGF vs CTR 
DataFrame with 57914 rows and 6 columns
                   baseMean log2FoldChange     lfcSE      stat     pvalue
                  &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;
ENSG00000000003 1430.562846      -0.256207 0.0813936  -3.14775 0.00164531
ENSG00000000005    0.113566       0.000000 4.0804559   0.00000 1.00000000
ENSG00000000419 1790.537536      -0.144906 0.1008195  -1.43728 0.15063862
ENSG00000000457  640.692302      -0.310636 0.1137555  -2.73073 0.00631933
ENSG00000000460  206.179026       0.160027 0.1567963   1.02061 0.30744046
...                     ...            ...       ...       ...        ...
ENSG00000284744    8.307038      -0.209531  0.800702 -0.261684   0.793565
ENSG00000284745    0.000000             NA        NA        NA         NA
ENSG00000284746    0.101097      -1.050150  4.080456 -0.257361   0.796900
ENSG00000284747   28.783710      -0.297530  0.448931 -0.662751   0.507490
ENSG00000284748    0.548323       0.873372  4.080456  0.214038   0.830517
                      padj
                 &lt;numeric&gt;
ENSG00000000003 0.00906995
ENSG00000000005         NA
ENSG00000000419 0.30807839
ENSG00000000457 0.02685529
ENSG00000000460 0.50350287
...                    ...
ENSG00000284744         NA
ENSG00000284745         NA
ENSG00000284746         NA
ENSG00000284747   0.689074
ENSG00000284748         NA</code></pre>
</div>
</div>
<p>IF we wanted to change the order we can do:-</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Changing the direction of the contrast</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(de_condition, <span class="at">contrast=</span><span class="fu">c</span>(<span class="st">"condition"</span>,<span class="st">"CTR"</span>,<span class="st">"TGF"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or indeed using the same syntax you can perform contrasts using the <code>IR</code> condition.</p>
<section id="exercise" class="level2">
<h2 class="anchored" data-anchor-id="exercise">Exercise</h2>
<ul>
<li>Re-run the analysis to find differentially-expressed genes between the <code>IR</code> treated samples and <code>CTR</code></li>
<li>Write a csv file that contains results for the genes that have an <em>adjusted</em> p-value less than 0.05 and a log2 fold change more than 1, or less than -1 in the contrast of TGF vs CTRL.</li>
<li>Use the <code>plotCounts</code> function to visually-inspect the most statistically-significant gene identified</li>
</ul>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Use the DESeq function once and once only!
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<p>You should perform the full differential expression analysis by calling DESeq(dds) only once on your combined dataset. Then, you use the results() function multiple times to extract the specific pairwise comparisons or complex contrasts you are interested in.</p>
</section>
</section>
<section id="more-complex-designs" class="level1">
<h1>More complex designs</h1>
<p>The examples we have used so far have performed a differential expression analysis using a named column in the <code>colData</code> object. The <code>DESeq2</code> package is capable of performing more complex analyses that can take multiple factors into consideration at the same time; so-called “multi-factor designs”</p>
<ul>
<li><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs">Multi-factor designs in DESeq2</a></li>
</ul>
<p>The use of such a design could be motivated by discovering sources of technical variation in our data that might obscure the biological differences we would like to compare. e.g.</p>
<p><img src="../..\files/training/bulk_rnaseq/batch_effect.png" class="img-fluid"></p>
<p>In the example image above the main source of variation is the batch in which the samples were sequenced. A multi-factor analysis to compare the various conditions, but “correct” for differences in batch, would be as follows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Don't run this. It's just a code example</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">design</span>(MY_DATA) <span class="ot">&lt;-</span> <span class="er">~</span> batch <span class="sc">+</span> condition</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Likewise, if we have different treatments applied to difference cell-lines, but the main source of variation is the cell line the following could be used.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Don't run this. It's just a code example</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">design</span>(MY_DATA) <span class="ot">&lt;-</span> <span class="er">~</span>cell_line <span class="sc">+</span> treatment</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="adding-annotation-to-the-deseq2-results" class="level1">
<h1>Adding annotation to the DESeq2 results</h1>
<p>We would love to share these results with our collaborators, or search for our favourite gene in the results. However, the results are not very useful in there current form as each row is named according to an <em>Ensembl</em> identifier. Whilst gene symbols are problematic and can change over time, they are the names that are most recognisable and make the results easier to navigate.</p>
<p>There are a number of ways to add annotation, but we will demonstrate how to do this using the <em>org.Hs.eg.db</em> package. This package is one of several <em>organism-level</em> packages in Bioconductor that are re-built every 6 months. These packages are listed on the <a href="http://bioconductor.org/packages/release/BiocViews.html#___AnnotationData">annotation section</a> of the Bioconductor, and are installed in the same way as regular Bioconductor packages. An alternative approach is to use <code>biomaRt</code>, an interface to the <a href="http://www.biomart.org/">BioMart</a> resource. BioMart is much more comprehensive, but the organism packages do not require online access once downloaded.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Only execute when you need to install the package</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"org.Hs.eg.db"</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="co"># For Human</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"org.Hs.eg.db"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The packages are larger in size that Bioconductor software packages, but essentially they are databases that can be used to make <em>offline</em> queries. An alternatve (<code>biomaRt</code>) that connects to the ensembl <a href="https://www.ensembl.org/info/data/biomart/index.html">biomart</a> resource will be discussed later.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>First we need to decide what information we want. In order to see what we can extract we can run the <code>columns</code> function on the annotation database.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">columns</span>(org.Hs.eg.db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"         
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"        
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
[26] "UNIPROT"     </code></pre>
</div>
</div>
<p>We are going to filter the database by a key or set of keys in order to extract the information we want. Valid names for the key can be retrieved with the <code>keytypes</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">keytypes</span>(org.Hs.eg.db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"         
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"        
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
[26] "UNIPROT"     </code></pre>
</div>
</div>
<p>We should see <code>ENSEMBL</code>, which is the type of key we are going to use in this case. If we are unsure what values are acceptable for the key, we can check what keys are valid with <code>keys</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">keys</span>(org.Hs.eg.db, <span class="at">keytype=</span><span class="st">"ENSEMBL"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "ENSG00000121410" "ENSG00000175899" "ENSG00000291190" "ENSG00000171428"
 [5] "ENSG00000156006" "ENSG00000196136" "ENSG00000114771" "ENSG00000127837"
 [9] "ENSG00000129673" "ENSG00000090861"</code></pre>
</div>
</div>
<p>For the top gene in our analysis the call to the function would be:-</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(org.Hs.eg.db, <span class="at">keys=</span>top_up,</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>,<span class="at">columns=</span><span class="fu">c</span>(<span class="st">"SYMBOL"</span>,<span class="st">"GENENAME"</span>)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Unfortunately, the authors of <code>dplyr</code> and <code>AnnotationDbi</code> have both decided to use the name <code>select</code> in their packages. To avoid confusion and errors, the following code is sometimes used:-</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>AnnotationDbi<span class="sc">::</span><span class="fu">select</span>(org.Hs.eg.db, <span class="at">keys=</span>top_up,<span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>,<span class="at">columns=</span><span class="fu">c</span>(<span class="st">"SYMBOL"</span>,<span class="st">"GENENAME"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>          ENSEMBL SYMBOL
1 ENSG00000119681  LTBP2
                                                  GENENAME
1 latent transforming growth factor beta binding protein 2</code></pre>
</div>
</div>
<p>To annotate our results, we definitely want gene symbols and perhaps the full gene name. Let’s build up our annotation information into a new data frame using the <code>select</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>anno <span class="ot">&lt;-</span> AnnotationDbi<span class="sc">::</span><span class="fu">select</span>(org.Hs.eg.db,<span class="at">keys=</span><span class="fu">rownames</span>(dds),</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">columns=</span><span class="fu">c</span>(<span class="st">"SYMBOL"</span>,<span class="st">"GENENAME"</span>),</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">keytype=</span><span class="st">"ENSEMBL"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Have a look at the annotation</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(anno)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          ENSEMBL SYMBOL
1 ENSG00000000003 TSPAN6
2 ENSG00000000005   TNMD
3 ENSG00000000419   DPM1
4 ENSG00000000457  SCYL3
5 ENSG00000000460  FIRRM
6 ENSG00000000938    FGR
                                                     GENENAME
1                                               tetraspanin 6
2                                                 tenomodulin
3 dolichyl-phosphate mannosyltransferase subunit 1, catalytic
4                                    SCY1 like pseudokinase 3
5   FIGNL1 interacting regulator of recombination and mitosis
6              FGR proto-oncogene, Src family tyrosine kinase</code></pre>
</div>
</div>
<p>However, we have a problem that the resulting data frame has more rows than our results table. This is due to the <em>one-to-many</em> relationships that often occur when mapping between various identifiers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(anno)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 58969     3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 57914     9</code></pre>
</div>
</div>
<p>Such duplicated entries can be identified using the <code>duplicated</code> function. Fortunately, there are not too many so hopefully we won’t lose too much information if we discard the entries that are duplicated. The first occurrence of the duplicated ID will still be included in the table.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>anno <span class="ot">&lt;-</span> AnnotationDbi<span class="sc">::</span><span class="fu">select</span>(org.Hs.eg.db,<span class="at">keys=</span><span class="fu">rownames</span>(dds),</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">columns=</span><span class="fu">c</span>(<span class="st">"ENSEMBL"</span>,<span class="st">"SYMBOL"</span>,<span class="st">"GENENAME"</span>,<span class="st">"ENTREZID"</span>),</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">keytype=</span><span class="st">"ENSEMBL"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">duplicated</span>(ENSEMBL))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(anno)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 57914     4</code></pre>
</div>
</div>
<p>We can bind in the annotation information to the <code>results</code> data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>results_annotated <span class="ot">&lt;-</span> <span class="fu">results</span>(de_condition,<span class="at">tidy=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(anno, <span class="at">by=</span><span class="fu">c</span>(<span class="st">"row"</span><span class="ot">=</span><span class="st">"ENSEMBL"</span>))</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(results_annotated)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              row     baseMean log2FoldChange      lfcSE       stat      pvalue
1 ENSG00000000003 1430.5628457     -0.2562067 0.08139356 -3.1477521 0.001645312
2 ENSG00000000005    0.1135661     -0.0883933 4.08045587 -0.0216626 0.982717094
3 ENSG00000000419 1790.5375358     -0.1449058 0.10081950 -1.4372794 0.150638620
4 ENSG00000000457  640.6923020     -0.3106360 0.11375545 -2.7307347 0.006319333
5 ENSG00000000460  206.1790264      0.1600274 0.15679627  1.0206074 0.307440460
6 ENSG00000000938    7.4126750     -1.9496601 0.94841767 -2.0556978 0.039811655
         padj SYMBOL
1 0.009069945 TSPAN6
2          NA   TNMD
3 0.308078390   DPM1
4 0.026855292  SCYL3
5 0.503502873  FIRRM
6          NA    FGR
                                                     GENENAME ENTREZID
1                                               tetraspanin 6     7105
2                                                 tenomodulin    64102
3 dolichyl-phosphate mannosyltransferase subunit 1, catalytic     8813
4                                    SCY1 like pseudokinase 3    57147
5   FIGNL1 interacting regulator of recombination and mitosis    55732
6              FGR proto-oncogene, Src family tyrosine kinase     2268</code></pre>
</div>
</div>
<p>We can save the results table using the <code>write.csv</code> function, which writes the results out to a csv file that you can open in excel.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(results_annotated,<span class="at">file=</span><span class="st">"de_analysis/condition_TGF_vs_CTR_DESeq_annotated.csv"</span>,<span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="exercise-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1">Exercise</h2>
<ul>
<li>The publication gives examples of <code>COL1A1</code>, <code>COL1A2</code> and <code>COL3A1</code> as genes that are <em>up-regulated</em> in TGF-treated samples vs controls (Figure 6C). Use your data to verify this by
<ol type="i">
<li>extracting their p-values</li>
<li>plotting the counts for these genes</li>
</ol></li>
</ul>
</section>
<section id="the-volcano-plot" class="level2">
<h2 class="anchored" data-anchor-id="the-volcano-plot">The volcano plot</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnhancedVolcano)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ggplot2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ggrepel</code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">EnhancedVolcano</span>(results_annotated, <span class="at">x =</span> <span class="st">"log2FoldChange"</span>, <span class="at">y =</span> <span class="st">"padj"</span>, <span class="at">lab =</span> results_annotated<span class="sc">$</span>SYMBOL)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.
ℹ The deprecated feature was likely used in the EnhancedVolcano package.
  Please report the issue to the authors.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `size` argument of `element_line()` is deprecated as of ggplot2 3.4.0.
ℹ Please use the `linewidth` argument instead.
ℹ The deprecated feature was likely used in the EnhancedVolcano package.
  Please report the issue to the authors.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="exporting-normalized-counts" class="level3">
<h3 class="anchored" data-anchor-id="exporting-normalized-counts">Exporting normalized counts</h3>
<p>The <code>DESeq</code> workflow applies <em>median of ratios normalization</em> that accounts for differences in sequencing depth between samples. The user does not usually need to run this step. However, if you want a matrix of counts for some application outside of Bioconductor the values can be extracted from the <code>dds</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds) </span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>countMatrix <span class="ot">&lt;-</span><span class="fu">counts</span>(dds, <span class="at">normalized=</span><span class="cn">TRUE</span>) </span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(countMatrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 1_CTR_BC_2  2_TGF_BC_4  3_IR_BC_5  4_CTR_BC_6   5_TGF_BC_7
ENSG00000000003 1496.753711 1309.582900 1503.06012 1550.428865 1326.3138781
ENSG00000000005    0.000000    0.000000    0.00000    0.000000    0.0000000
ENSG00000000419 1681.596633 1502.591885 2089.94798 1645.965855 1826.4150250
ENSG00000000457  661.642869  522.309405  673.12901  666.939177  629.4048655
ENSG00000000460  233.186455  261.577968  129.92174  203.812245  242.4444723
ENSG00000000938    9.479124    4.232653   12.32016    5.459257    0.9507626
                 6_IR_BC_12 7_CTR_BC_13 8_TGF_BC_14  9_IR_BC_15
ENSG00000000003 1465.288919  1639.42512 1287.173576 1297.038521
ENSG00000000005    0.000000     0.00000    0.000000    1.022095
ENSG00000000419 2058.823670  1893.34636 1392.402365 2023.748047
ENSG00000000457  622.515941   791.26495  557.148855  641.875643
ENSG00000000460  125.198737   258.13570  271.527857  129.806062
ENSG00000000938    2.318495    20.01869    3.758171    8.176760</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(countMatrix,<span class="at">file=</span><span class="st">"normalized_counts.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="full-deseq-workflow" class="level1">
<h1>Full DESeq workflow</h1>
<p>The median of ratios normalisation method is employed in DESeq2 to account for <em>sequencing depth</em> and <em>RNA composition</em>. Let’s go through a short worked example (courtesy of <a href="here">https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html</a>) to explain the process.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="do">## create a small example matrix of "counts"</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>test_data <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1489</span>,<span class="dv">22</span>,<span class="dv">793</span>,<span class="dv">76</span>,<span class="dv">521</span>,<span class="dv">906</span>,<span class="dv">13</span>,<span class="dv">410</span>,<span class="dv">42</span>,<span class="dv">1196</span>),<span class="at">nrow=</span><span class="dv">5</span>)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(test_data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"EF2A"</span>,<span class="st">"ABCD1"</span>,<span class="st">"MEFV"</span>,<span class="st">"BAG1"</span>,<span class="st">"MOV10"</span>)</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(test_data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"SampleA"</span>,<span class="st">"SampleB"</span>)</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>test_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      SampleA SampleB
EF2A     1489     906
ABCD1      22      13
MEFV      793     410
BAG1       76      42
MOV10     521    1196</code></pre>
</div>
</div>
<p>Firstly, an “average” or reference sample is created that represents the counts on a typical sample in the dataset. The <em>geometric mean</em> is used rather than the <em>arithmetic</em> mean. In other words the individual counts are multiplied rather than summed and the measure should be more robust to outliers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>psuedo_ref <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">rowProds</span>(test_data))</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>psuedo_ref</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      EF2A      ABCD1       MEFV       BAG1      MOV10 
1161.47923   16.91153  570.20172   56.49779  789.37697 </code></pre>
</div>
</div>
<p>A ratios of sample to “psuedo reference” are then calculated for each gene. We are assuming that most genes are not changing dramatically, so this ratio should be somewhere around 1.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>test_data<span class="sc">/</span>psuedo_ref</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        SampleA   SampleB
EF2A  1.2819859 0.7800398
ABCD1 1.3008873 0.7687061
MEFV  1.3907359 0.7190438
BAG1  1.3451854 0.7433919
MOV10 0.6600142 1.5151189</code></pre>
</div>
</div>
<p><code>DESeq2</code> defines size factors as being the <em>median</em> of these ratios for each sample (median is used so any outlier genes will not affect the normalisation).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>norm_factors <span class="ot">&lt;-</span> <span class="fu">colMedians</span>(test_data<span class="sc">/</span>psuedo_ref)</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>norm_factors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  SampleA   SampleB 
1.3008873 0.7687061 </code></pre>
</div>
</div>
<p>Individual samples can then normalised by dividing the count for each gene by the corresponding normalization factor.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>test_data[,<span class="dv">1</span>] <span class="sc">/</span> norm_factors[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      EF2A      ABCD1       MEFV       BAG1      MOV10 
1144.60340   16.91153  609.58395   58.42166  400.49589 </code></pre>
</div>
</div>
<p>and for the second sample…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>test_data[,<span class="dv">2</span>] <span class="sc">/</span> norm_factors[<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      EF2A      ABCD1       MEFV       BAG1      MOV10 
1178.60387   16.91153  533.36378   54.63727 1555.86118 </code></pre>
</div>
</div>
<p>The size factors for each sample in our dataset can be calculated using the <code>estimateSizeFactorsForMatrix</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>sf <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactorsForMatrix</span>(<span class="fu">assay</span>(dds))</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>sf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 1_CTR_BC_2  2_TGF_BC_4   3_IR_BC_5  4_CTR_BC_6  5_TGF_BC_7  6_IR_BC_12 
  1.0549498   1.1812922   0.8928452   1.0990507   1.0517872   0.8626285 
7_CTR_BC_13 8_TGF_BC_14  9_IR_BC_15 
  0.9491132   1.0643475   0.9783827 </code></pre>
</div>
</div>
<p>The estimation of these factors can also take gene-lengths into account, and this is implemented in the <code>estimateSizeFactors</code> function. Extra normalization factor data is added to the <code>dds</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds)</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In preparation for differential expression DESeq2 also need a reliable estimate of the variability of each gene; which it calls <em>dispersion</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateDispersions</span>(dds)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A statistical test can then be applied. As the data are count-based and not normally-distributed a t-test would not be appropriate. Most tests are based on a <em>Poisson</em> or <em>negative-binomial</em> distribution; negative binomial in the case of <code>DESeq2</code>. Although you might not be familiar with the negative binomial, the results should be in a familiar form with fold-changes and p-values for each gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">nbinomWaldTest</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It may seem like there is a lot to remember, but fortunately there is one convenient function that will apply the three steps (<code>DESeq</code>). The messages printed serve as reminders of the steps included.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>